// zxing-wasm/dist/es/writer/index.js
import { A as e, C as t, D as n, E as r, F as i, I as a, L as o, M as s, N as c, O as l, P as u, S as d, T as f, a as p, b as m, c as h, g, h as _, i as ee, j as te, k as ne, o as re, r as ie, t as ae, w as oe, x as v, y } from "../share.js";
//#region src/writer/zxing_writer.js
async function b(e = {}) {
	var t, n, r, i = e, a = !!globalThis.window, o = typeof Bun < "u", s = !!globalThis.WorkerGlobalScope;
	!((n = globalThis.process) == null || (n = n.versions) == null) && n.node && ((r = globalThis.process) == null || r.type);
	var c = "./this.program", l, u = "";
	function d(e) {
		return i.locateFile ? i.locateFile(e, u) : u + e;
	}
	var f, p;
	if (a || s || o) {
		try {
			u = new URL(".", l).href;
		} catch {}
		s && (p = (e) => {
			var t = new XMLHttpRequest();
			return t.open("GET", e, !1), t.responseType = "arraybuffer", t.send(null), new Uint8Array(t.response);
		}), f = async (e) => {
			var t = await fetch(e, { credentials: "same-origin" });
			if (t.ok) return t.arrayBuffer();
			throw Error(t.status + " : " + t.url);
		};
	}
	var m = console.log.bind(console), h = console.error.bind(console), g, _ = !1, ee, te, ne = !1;
	function re() {
		var e = Q.buffer;
		w = new Int8Array(e), S = new Int16Array(e), i.HEAPU8 = D = new Uint8Array(e), T = new Uint16Array(e), C = new Int32Array(e), E = new Uint32Array(e), fe = new Float32Array(e), pe = new Float64Array(e);
	}
	function ie() {
		if (i.preRun) for (typeof i.preRun == "function" && (i.preRun = [i.preRun]); i.preRun.length;) ve(i.preRun.shift());
		me(_e);
	}
	function ae() {
		ne = !0, $.oa();
	}
	function oe() {
		if (i.postRun) for (typeof i.postRun == "function" && (i.postRun = [i.postRun]); i.postRun.length;) ge(i.postRun.shift());
		me(he);
	}
	function v(e) {
		var t, n;
		(t = i.onAbort) == null || t.call(i, e), e = "Aborted(" + e + ")", h(e), _ = !0, e += ". Build with -sASSERTIONS for more info.";
		var r = new WebAssembly.RuntimeError(e);
		throw (n = te) == null || n(r), r;
	}
	var y;
	function b() {
		return d("zxing_writer.wasm");
	}
	function x(e) {
		if (e == y && g) return new Uint8Array(g);
		if (p) return p(e);
		throw "both async and sync fetching of the wasm failed";
	}
	async function se(e) {
		if (!g) try {
			var t = await f(e);
			return new Uint8Array(t);
		} catch {}
		return x(e);
	}
	async function ce(e, t) {
		try {
			var n = await se(e);
			return await WebAssembly.instantiate(n, t);
		} catch (e) {
			h(`failed to asynchronously prepare wasm: ${e}`), v(e);
		}
	}
	async function le(e, t, n) {
		if (!e && WebAssembly.instantiateStreaming) try {
			var r = fetch(t, { credentials: "same-origin" });
			return await WebAssembly.instantiateStreaming(r, n);
		} catch (e) {
			h(`wasm streaming compile failed: ${e}`), h("falling back to ArrayBuffer instantiation");
		}
		return ce(t, n);
	}
	function ue() {
		return { a: Sn };
	}
	async function de() {
		function e(e, t) {
			return $ = e.exports, xn($), re(), $;
		}
		function t(t) {
			return e(t.instance);
		}
		var n = ue();
		return i.instantiateWasm ? new Promise((t, r) => {
			i.instantiateWasm(n, (n, r) => {
				t(e(n, r));
			});
		}) : (y != null || (y = b()), t(await le(g, y, n)));
	}
	var S, C, w, fe, pe, T, E, D, me = (e) => {
		for (; e.length > 0;) e.shift()(i);
	}, he = [], ge = (e) => he.push(e), _e = [], ve = (e) => _e.push(e), O = (e) => pn(e), k = () => mn(), A = [], j = 0, ye = (e) => {
		var t = new xe(e);
		return t.get_caught() || (t.set_caught(!0), j--), t.set_rethrown(!1), A.push(t), vn(e);
	}, M = 0, be = () => {
		Z(0, 0);
		var e = A.pop();
		hn(e.excPtr), M = 0;
	};
	class xe {
		constructor(e) {
			this.excPtr = e, this.ptr = e - 24;
		}
		set_type(e) {
			E[this.ptr + 4 >> 2] = e;
		}
		get_type() {
			return E[this.ptr + 4 >> 2];
		}
		set_destructor(e) {
			E[this.ptr + 8 >> 2] = e;
		}
		get_destructor() {
			return E[this.ptr + 8 >> 2];
		}
		set_caught(e) {
			e = +!!e, w[this.ptr + 12] = e;
		}
		get_caught() {
			return w[this.ptr + 12] != 0;
		}
		set_rethrown(e) {
			e = +!!e, w[this.ptr + 13] = e;
		}
		get_rethrown() {
			return w[this.ptr + 13] != 0;
		}
		init(e, t) {
			this.set_adjusted_ptr(0), this.set_type(e), this.set_destructor(t);
		}
		set_adjusted_ptr(e) {
			E[this.ptr + 16 >> 2] = e;
		}
		get_adjusted_ptr() {
			return E[this.ptr + 16 >> 2];
		}
	}
	var N = (e) => fn(e), P = (e) => {
		var t = M;
		if (!t) return N(0), 0;
		var n = new xe(t);
		n.set_adjusted_ptr(t);
		var r = n.get_type();
		if (!r) return N(0), t;
		for (var i of e) {
			if (i === 0 || i === r) break;
			var a = n.ptr + 16;
			if (_n(i, r, a)) return N(i), t;
		}
		return N(r), t;
	}, Se = () => P([]), Ce = (e) => P([e]), we = (e, t) => P([e, t]), Te = () => {
		var e = A.pop();
		e || v("no exception to throw");
		var t = e.excPtr;
		throw e.get_rethrown() || (A.push(e), e.set_rethrown(!0), e.set_caught(!1), j++), gn(t), M = t, M;
	}, Ee = (e, t, n) => {
		throw new xe(e).init(t, n), gn(e), M = e, j++, M;
	}, De = () => j, Oe = (e) => {
		throw M || (M = e), M;
	}, ke = globalThis.TextDecoder && new TextDecoder(), Ae = (e, t, n, r) => {
		var i = t + n;
		if (r) return i;
		for (; e[t] && !(t >= i);) ++t;
		return t;
	}, je = function(e) {
		let t = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : 0, n = arguments.length > 2 ? arguments[2] : void 0, r = arguments.length > 3 ? arguments[3] : void 0;
		var i = Ae(e, t, n, r);
		if (i - t > 16 && e.buffer && ke) return ke.decode(e.subarray(t, i));
		for (var a = ""; t < i;) {
			var o = e[t++];
			if (!(o & 128)) {
				a += String.fromCharCode(o);
				continue;
			}
			var s = e[t++] & 63;
			if ((o & 224) == 192) {
				a += String.fromCharCode((o & 31) << 6 | s);
				continue;
			}
			var c = e[t++] & 63;
			if (o = (o & 240) == 224 ? (o & 15) << 12 | s << 6 | c : (o & 7) << 18 | s << 12 | c << 6 | e[t++] & 63, o < 65536) a += String.fromCharCode(o);
			else {
				var l = o - 65536;
				a += String.fromCharCode(55296 | l >> 10, 56320 | l & 1023);
			}
		}
		return a;
	}, Me = (e, t, n) => e ? je(D, e, t, n) : "", Ne = {
		varargs: void 0,
		getStr(e) {
			return Me(e);
		}
	};
	function Pe(e, t, n) {
		return Ne.varargs = n, 0;
	}
	function Fe(e, t, n) {
		return Ne.varargs = n, 0;
	}
	var Ie = (e, t, n) => {};
	function Le(e, t, n, r) {
		Ne.varargs = r;
	}
	var Re = (e, t) => {}, ze = () => v(""), F = {}, Be = (e) => {
		for (; e.length;) {
			var t = e.pop();
			e.pop()(t);
		}
	};
	function I(e) {
		return this.fromWireType(E[e >> 2]);
	}
	var L = {}, R = {}, z = {}, Ve = class extends Error {
		constructor(e) {
			super(e), this.name = "InternalError";
		}
	}, He = (e) => {
		throw new Ve(e);
	}, Ue = (e, t, n) => {
		e.forEach((e) => z[e] = t);
		function r(t) {
			var r = n(t);
			r.length !== e.length && He("Mismatched type converter count");
			for (var i = 0; i < e.length; ++i) H(e[i], r[i]);
		}
		var i = Array(t.length), a = [], o = 0;
		{
			let e = t;
			for (let t = 0; t < e.length; ++t) {
				let n = e[t];
				R.hasOwnProperty(n) ? i[t] = R[n] : (a.push(n), L.hasOwnProperty(n) || (L[n] = []), L[n].push(() => {
					i[t] = R[n], ++o, o === a.length && r(i);
				}));
			}
		}
		a.length === 0 && r(i);
	}, We = (e) => {
		var t = F[e];
		delete F[e];
		var n = t.rawConstructor, r = t.rawDestructor, i = t.fields, a = i.map((e) => e.getterReturnType).concat(i.map((e) => e.setterArgumentType));
		Ue([e], a, (e) => {
			var a = {};
			{
				let t = i;
				for (let n = 0; n < t.length; ++n) {
					let r = t[n], o = e[n], s = r.getter, c = r.getterContext, l = e[n + i.length], u = r.setter, d = r.setterContext;
					a[r.fieldName] = {
						read: (e) => o.fromWireType(s(c, e)),
						write: (e, t) => {
							var n = [];
							u(d, e, l.toWireType(n, t)), Be(n);
						},
						optional: o.optional
					};
				}
			}
			return [{
				name: t.name,
				fromWireType: (e) => {
					var t = {};
					for (var n in a) t[n] = a[n].read(e);
					return r(e), t;
				},
				toWireType: (e, t) => {
					for (var i in a) if (!(i in t) && !a[i].optional) throw TypeError(`Missing field: "${i}"`);
					var o = n();
					for (i in a) a[i].write(o, t[i]);
					return e !== null && e.push(r, o), o;
				},
				readValueFromPointer: I,
				destructorFunction: r
			}];
		});
	}, Ge = (e, t, n, r, i) => {}, B = (e) => {
		for (var t = "";;) {
			var n = D[e++];
			if (!n) return t;
			t += String.fromCharCode(n);
		}
	}, Ke = class extends Error {
		constructor(e) {
			super(e), this.name = "BindingError";
		}
	}, V = (e) => {
		throw new Ke(e);
	};
	function qe(e, t) {
		let n = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {};
		var r = t.name;
		if (e || V(`type "${r}" must have a positive integer typeid pointer`), R.hasOwnProperty(e)) {
			if (n.ignoreDuplicateRegistrations) return;
			V(`Cannot register type '${r}' twice`);
		}
		if (R[e] = t, delete z[e], L.hasOwnProperty(e)) {
			var i = L[e];
			delete L[e], i.forEach((e) => e());
		}
	}
	function H(e, t) {
		return qe(e, t, arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {});
	}
	var Je = (e, t, n, r) => {
		t = B(t), H(e, {
			name: t,
			fromWireType: function(e) {
				return !!e;
			},
			toWireType: function(e, t) {
				return t ? n : r;
			},
			readValueFromPointer: function(e) {
				return this.fromWireType(D[e]);
			},
			destructorFunction: null
		});
	}, Ye = [], U = [
		0,
		1,
		,
		1,
		null,
		1,
		!0,
		1,
		!1,
		1
	], Xe = (e) => {
		e > 9 && --U[e + 1] === 0 && (U[e] = void 0, Ye.push(e));
	}, W = {
		toValue: (e) => (e || V(`Cannot use deleted val. handle = ${e}`), U[e]),
		toHandle: (e) => {
			switch (e) {
				case void 0: return 2;
				case null: return 4;
				case !0: return 6;
				case !1: return 8;
				default: {
					let t = Ye.pop() || U.length;
					return U[t] = e, U[t + 1] = 1, t;
				}
			}
		}
	}, Ze = {
		name: "emscripten::val",
		fromWireType: (e) => {
			var t = W.toValue(e);
			return Xe(e), t;
		},
		toWireType: (e, t) => W.toHandle(t),
		readValueFromPointer: I,
		destructorFunction: null
	}, Qe = (e) => H(e, Ze), $e = (e, t) => {
		switch (t) {
			case 4: return function(e) {
				return this.fromWireType(fe[e >> 2]);
			};
			case 8: return function(e) {
				return this.fromWireType(pe[e >> 3]);
			};
			default: throw TypeError(`invalid float width (${t}): ${e}`);
		}
	}, et = (e, t, n) => {
		t = B(t), H(e, {
			name: t,
			fromWireType: (e) => e,
			toWireType: (e, t) => t,
			readValueFromPointer: $e(t, n),
			destructorFunction: null
		});
	}, tt = (e, t) => Object.defineProperty(t, "name", { value: e });
	function nt(e) {
		for (var t = 1; t < e.length; ++t) if (e[t] !== null && e[t].destructorFunction === void 0) return !0;
		return !1;
	}
	function rt(e, t, n, r, i, a) {
		var o = t.length;
		o < 2 && V("argTypes array size mismatch! Must at least get return value and 'this' types!");
		var s = t[1] !== null && n !== null, c = nt(t), l = !t[0].isVoid, u = o - 2, d = Array(u), f = [], p = [];
		return tt(e, function() {
			p.length = 0;
			var e;
			f.length = s ? 2 : 1, f[0] = i, s && (e = t[1].toWireType(p, this), f[1] = e);
			for (var n = 0; n < u; ++n) d[n] = t[n + 2].toWireType(p, n < 0 || arguments.length <= n ? void 0 : arguments[n]), f.push(d[n]);
			var a = r(...f);
			function o(n) {
				if (c) Be(p);
				else for (var r = s ? 1 : 2; r < t.length; r++) {
					var i = r === 1 ? e : d[r - 2];
					t[r].destructorFunction !== null && t[r].destructorFunction(i);
				}
				if (l) return t[0].fromWireType(n);
			}
			return o(a);
		});
	}
	var it = (e, t, n) => {
		if (e[t].overloadTable === void 0) {
			var r = e[t];
			e[t] = function() {
				var r = [...arguments];
				return e[t].overloadTable.hasOwnProperty(r.length) || V(`Function '${n}' called with an invalid number of arguments (${r.length}) - expects one of (${e[t].overloadTable})!`), e[t].overloadTable[r.length].apply(this, r);
			}, e[t].overloadTable = [], e[t].overloadTable[r.argCount] = r;
		}
	}, at = (e, t, n) => {
		i.hasOwnProperty(e) ? ((n === void 0 || i[e].overloadTable !== void 0 && i[e].overloadTable[n] !== void 0) && V(`Cannot register public name '${e}' twice`), it(i, e, e), i[e].overloadTable.hasOwnProperty(n) && V(`Cannot register multiple overloads of a function with the same number of arguments (${n})!`), i[e].overloadTable[n] = t) : (i[e] = t, i[e].argCount = n);
	}, ot = (e, t) => {
		for (var n = [], r = 0; r < e; r++) n.push(E[t + r * 4 >> 2]);
		return n;
	}, st = (e, t, n) => {
		i.hasOwnProperty(e) || He("Replacing nonexistent public symbol"), i[e].overloadTable !== void 0 && n !== void 0 ? i[e].overloadTable[n] = t : (i[e] = t, i[e].argCount = n);
	}, G = {}, ct = (e, t, n) => {
		e = e.replace(/p/g, "i");
		var r = G[e];
		return r(t, ...n);
	}, lt = [], K = (e) => {
		var t = lt[e];
		return t || (lt[e] = t = bn.get(e)), t;
	}, ut = function(e, t) {
		let n = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : [];
		if (arguments.length > 3 && arguments[3] !== void 0 && arguments[3], e.includes("j")) return ct(e, t, n);
		var r = K(t)(...n);
		function i(e) {
			return e;
		}
		return i(r);
	}, dt = function(e, t) {
		let n = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : !1;
		return function() {
			return ut(e, t, [...arguments], n);
		};
	}, q = function(e, t) {
		arguments.length > 2 && arguments[2] !== void 0 && arguments[2], e = B(e);
		function n() {
			return e.includes("j") ? dt(e, t) : K(t);
		}
		var r = n();
		return typeof r != "function" && V(`unknown function pointer with signature ${e}: ${t}`), r;
	};
	class ft extends Error {}
	var pt = (e) => {
		var t = un(e), n = B(t);
		return X(t), n;
	}, mt = (e, t) => {
		var n = [], r = {};
		function i(e) {
			if (!r[e] && !R[e]) {
				if (z[e]) {
					z[e].forEach(i);
					return;
				}
				n.push(e), r[e] = !0;
			}
		}
		throw t.forEach(i), new ft(`${e}: ` + n.map(pt).join([", "]));
	}, ht = (e) => {
		e = e.trim();
		let t = e.indexOf("(");
		return t === -1 ? e : e.slice(0, t);
	}, gt = (e, t, n, r, i, a, o, s) => {
		var c = ot(t, n);
		e = B(e), e = ht(e), i = q(r, i, o), at(e, function() {
			mt(`Cannot call ${e} due to unbound types`, c);
		}, t - 1), Ue([], c, (n) => {
			var r = [n[0], null].concat(n.slice(1));
			return st(e, rt(e, r, null, i, a, o), t - 1), [];
		});
	}, _t = (e, t, n) => {
		switch (t) {
			case 1: return n ? (e) => w[e] : (e) => D[e];
			case 2: return n ? (e) => S[e >> 1] : (e) => T[e >> 1];
			case 4: return n ? (e) => C[e >> 2] : (e) => E[e >> 2];
			default: throw TypeError(`invalid integer width (${t}): ${e}`);
		}
	}, vt = (e, t, n, r, i) => {
		t = B(t);
		let a = r === 0, o = (e) => e;
		if (a) {
			var s = 32 - 8 * n;
			o = (e) => e << s >>> s, i = o(i);
		}
		H(e, {
			name: t,
			fromWireType: o,
			toWireType: (e, t) => t,
			readValueFromPointer: _t(t, n, r !== 0),
			destructorFunction: null
		});
	}, yt = (e, t, n) => {
		var r = [
			Int8Array,
			Uint8Array,
			Int16Array,
			Uint16Array,
			Int32Array,
			Uint32Array,
			Float32Array,
			Float64Array
		][t];
		function i(e) {
			var t = E[e >> 2], n = E[e + 4 >> 2];
			return new r(w.buffer, n, t);
		}
		n = B(n), H(e, {
			name: n,
			fromWireType: i,
			readValueFromPointer: i
		}, { ignoreDuplicateRegistrations: !0 });
	}, bt = (e, t, n, r) => {
		if (!(r > 0)) return 0;
		for (var i = n, a = n + r - 1, o = 0; o < e.length; ++o) {
			var s = e.codePointAt(o);
			if (s <= 127) {
				if (n >= a) break;
				t[n++] = s;
			} else if (s <= 2047) {
				if (n + 1 >= a) break;
				t[n++] = 192 | s >> 6, t[n++] = 128 | s & 63;
			} else if (s <= 65535) {
				if (n + 2 >= a) break;
				t[n++] = 224 | s >> 12, t[n++] = 128 | s >> 6 & 63, t[n++] = 128 | s & 63;
			} else {
				if (n + 3 >= a) break;
				t[n++] = 240 | s >> 18, t[n++] = 128 | s >> 12 & 63, t[n++] = 128 | s >> 6 & 63, t[n++] = 128 | s & 63, o++;
			}
		}
		return t[n] = 0, n - i;
	}, J = (e, t, n) => bt(e, D, t, n), xt = (e) => {
		for (var t = 0, n = 0; n < e.length; ++n) {
			var r = e.charCodeAt(n);
			r <= 127 ? t++ : r <= 2047 ? t += 2 : r >= 55296 && r <= 57343 ? (t += 4, ++n) : t += 3;
		}
		return t;
	}, St = (e, t) => {
		t = B(t);
		var n = !0;
		H(e, {
			name: t,
			fromWireType(e) {
				var t = E[e >> 2], r = e + 4, i;
				if (n) i = Me(r, t, !0);
				else {
					i = "";
					for (var a = 0; a < t; ++a) i += String.fromCharCode(D[r + a]);
				}
				return X(e), i;
			},
			toWireType(e, t) {
				t instanceof ArrayBuffer && (t = new Uint8Array(t));
				var r, i = typeof t == "string";
				i || ArrayBuffer.isView(t) && t.BYTES_PER_ELEMENT == 1 || V("Cannot pass non-string to std::string"), r = n && i ? xt(t) : t.length;
				var a = dn(4 + r + 1), o = a + 4;
				if (E[a >> 2] = r, i) if (n) J(t, o, r + 1);
				else for (var s = 0; s < r; ++s) {
					var c = t.charCodeAt(s);
					c > 255 && (X(a), V("String has UTF-16 code units that do not fit in 8 bits")), D[o + s] = c;
				}
				else D.set(t, o);
				return e !== null && e.push(X, a), a;
			},
			readValueFromPointer: I,
			destructorFunction(e) {
				X(e);
			}
		});
	}, Ct = globalThis.TextDecoder ? new TextDecoder("utf-16le") : void 0, wt = (e, t, n) => {
		var r = e >> 1, i = Ae(T, r, t / 2, n);
		if (i - r > 16 && Ct) return Ct.decode(T.subarray(r, i));
		for (var a = "", o = r; o < i; ++o) {
			var s = T[o];
			a += String.fromCharCode(s);
		}
		return a;
	}, Tt = (e, t, n) => {
		if (n != null || (n = 2147483647), n < 2) return 0;
		n -= 2;
		for (var r = t, i = n < e.length * 2 ? n / 2 : e.length, a = 0; a < i; ++a) {
			var o = e.charCodeAt(a);
			S[t >> 1] = o, t += 2;
		}
		return S[t >> 1] = 0, t - r;
	}, Et = (e) => e.length * 2, Dt = (e, t, n) => {
		for (var r = "", i = e >> 2, a = 0; !(a >= t / 4); a++) {
			var o = E[i + a];
			if (!o && !n) break;
			r += String.fromCodePoint(o);
		}
		return r;
	}, Ot = (e, t, n) => {
		if (n != null || (n = 2147483647), n < 4) return 0;
		for (var r = t, i = r + n - 4, a = 0; a < e.length; ++a) {
			var o = e.codePointAt(a);
			if (o > 65535 && a++, C[t >> 2] = o, t += 4, t + 4 > i) break;
		}
		return C[t >> 2] = 0, t - r;
	}, kt = (e) => {
		for (var t = 0, n = 0; n < e.length; ++n) e.codePointAt(n) > 65535 && n++, t += 4;
		return t;
	}, At = (e, t, n) => {
		n = B(n);
		var r, i, a;
		t === 2 ? (r = wt, i = Tt, a = Et) : (r = Dt, i = Ot, a = kt), H(e, {
			name: n,
			fromWireType: (e) => {
				var n = E[e >> 2], i = r(e + 4, n * t, !0);
				return X(e), i;
			},
			toWireType: (e, r) => {
				typeof r != "string" && V(`Cannot pass non-string to C++ string type ${n}`);
				var o = a(r), s = dn(4 + o + t);
				return E[s >> 2] = o / t, i(r, s + 4, o + t), e !== null && e.push(X, s), s;
			},
			readValueFromPointer: I,
			destructorFunction(e) {
				X(e);
			}
		});
	}, jt = (e, t, n, r, i, a) => {
		F[e] = {
			name: B(t),
			rawConstructor: q(n, r),
			rawDestructor: q(i, a),
			fields: []
		};
	}, Mt = (e, t, n, r, i, a, o, s, c, l) => {
		F[e].fields.push({
			fieldName: B(t),
			getterReturnType: n,
			getter: q(r, i),
			getterContext: a,
			setterArgumentType: o,
			setter: q(s, c),
			setterContext: l
		});
	}, Nt = (e, t) => {
		t = B(t), H(e, {
			isVoid: !0,
			name: t,
			fromWireType: () => void 0,
			toWireType: (e, t) => void 0
		});
	}, Pt = [], Ft = (e) => {
		var t = Pt.length;
		return Pt.push(e), t;
	}, It = (e, t) => {
		var n = R[e];
		return n === void 0 && V(`${t} has unknown type ${pt(e)}`), n;
	}, Lt = (e, t) => {
		for (var n = Array(e), r = 0; r < e; ++r) n[r] = It(E[t + r * 4 >> 2], `parameter ${r}`);
		return n;
	}, Rt = (e, t, n) => {
		var r = [], i = e(r, n);
		return r.length && (E[t >> 2] = W.toHandle(r)), i;
	}, zt = {}, Bt = (e) => {
		var t = zt[e];
		return t === void 0 ? B(e) : t;
	}, Vt = (e, t, n) => {
		var r = 8, [i, ...a] = Lt(e, t), o = i.toWireType.bind(i), s = a.map((e) => e.readValueFromPointer.bind(e));
		e--;
		var c = Array(e);
		return Ft(tt(`methodCaller<(${a.map((e) => e.name)}) => ${i.name}>`, (t, i, a, l) => {
			for (var u = 0, d = 0; d < e; ++d) c[d] = s[d](l + u), u += r;
			var f;
			switch (n) {
				case 0:
					f = W.toValue(t).apply(null, c);
					break;
				case 2:
					f = Reflect.construct(W.toValue(t), c);
					break;
				case 3:
					f = c[0];
					break;
				case 1:
					f = W.toValue(t)[Bt(i)](...c);
					break;
			}
			return Rt(o, a, f);
		}));
	}, Ht = (e) => e ? (e = Bt(e), W.toHandle(globalThis[e])) : W.toHandle(globalThis), Ut = (e) => {
		e > 9 && (U[e + 1] += 1);
	}, Wt = (e, t, n, r, i) => Pt[e](t, n, r, i), Gt = (e) => {
		Be(W.toValue(e)), Xe(e);
	}, Kt = (e, t, n, r) => {
		var i = (/* @__PURE__ */ new Date()).getFullYear(), a = new Date(i, 0, 1), o = new Date(i, 6, 1), s = a.getTimezoneOffset(), c = o.getTimezoneOffset(), l = Math.max(s, c);
		E[e >> 2] = l * 60, C[t >> 2] = Number(s != c);
		var u = (e) => {
			var t = e >= 0 ? "-" : "+", n = Math.abs(e);
			return `UTC${t}${String(Math.floor(n / 60)).padStart(2, "0")}${String(n % 60).padStart(2, "0")}`;
		}, d = u(s), f = u(c);
		c < s ? (J(d, n, 17), J(f, r, 17)) : (J(d, r, 17), J(f, n, 17));
	}, qt = () => 2147483648, Jt = (e, t) => Math.ceil(e / t) * t, Yt = (e) => {
		var t = (e - Q.buffer.byteLength + 65535) / 65536 | 0;
		try {
			return Q.grow(t), re(), 1;
		} catch {}
	}, Xt = (e) => {
		var t = D.length;
		e >>>= 0;
		var n = qt();
		if (e > n) return !1;
		for (var r = 1; r <= 4; r *= 2) {
			var i = t * (1 + .2 / r);
			if (i = Math.min(i, e + 100663296), Yt(Math.min(n, Jt(Math.max(e, i), 65536)))) return !0;
		}
		return !1;
	}, Zt = {}, Qt = () => c || "./this.program", Y = () => {
		if (!Y.strings) {
			var e, t, n = {
				USER: "web_user",
				LOGNAME: "web_user",
				PATH: "/",
				PWD: "/",
				HOME: "/home/web_user",
				LANG: ((e = (t = globalThis.navigator) == null ? void 0 : t.language) == null ? "C" : e).replace("-", "_") + ".UTF-8",
				_: Qt()
			};
			for (var r in Zt) Zt[r] === void 0 ? delete n[r] : n[r] = Zt[r];
			var i = [];
			for (var r in n) i.push(`${r}=${n[r]}`);
			Y.strings = i;
		}
		return Y.strings;
	}, $t = (e, t) => {
		var n = 0, r = 0;
		for (var i of Y()) {
			var a = t + n;
			E[e + r >> 2] = a, n += J(i, a, Infinity) + 1, r += 4;
		}
		return 0;
	}, en = (e, t) => {
		var n = Y();
		E[e >> 2] = n.length;
		var r = 0;
		for (var i of n) r += xt(i) + 1;
		return E[t >> 2] = r, 0;
	}, tn = (e) => 52, nn = (e, t, n, r) => 52, rn = (e, t) => t + 2097152 >>> 0 < 4194305 - !!e ? (e >>> 0) + t * 4294967296 : NaN;
	function an(e, t, n, r, i) {
		return rn(t, n), 70;
	}
	var on = [
		null,
		[],
		[]
	], sn = (e, t) => {
		var n = on[e];
		t === 0 || t === 10 ? ((e === 1 ? m : h)(je(n)), n.length = 0) : n.push(t);
	}, cn = (e, t, n, r) => {
		for (var i = 0, a = 0; a < n; a++) {
			var o = E[t >> 2], s = E[t + 4 >> 2];
			t += 8;
			for (var c = 0; c < s; c++) sn(e, D[o + c]);
			i += s;
		}
		return E[r >> 2] = i, 0;
	}, ln = (e) => e;
	if (i.noExitRuntime && i.noExitRuntime, i.print && (m = i.print), i.printErr && (h = i.printErr), i.wasmBinary && (g = i.wasmBinary), i.arguments && i.arguments, i.thisProgram && (c = i.thisProgram), i.preInit) for (typeof i.preInit == "function" && (i.preInit = [i.preInit]); i.preInit.length > 0;) i.preInit.shift()();
	var un, dn, X, Z, fn, pn, mn, hn, gn, _n, vn, yn, Q, bn;
	function xn(e) {
		un = e.pa, dn = i._malloc = e.ra, X = i._free = e.sa, Z = e.ta, fn = e.ua, pn = e.va, mn = e.wa, hn = e.xa, gn = e.ya, _n = e.za, vn = e.Aa, G.jiji = e.Ba, G.viijii = e.Ca, yn = G.jiiii = e.Da, G.iiiiij = e.Ea, G.iiiiijj = e.Fa, G.iiiiiijj = e.Ga, Q = e.na, bn = e.qa;
	}
	var Sn = {
		t: ye,
		u: be,
		a: Se,
		g: Ce,
		v: we,
		_: Te,
		p: Ee,
		Z: De,
		e: Oe,
		L: Pe,
		da: Fe,
		ba: Ie,
		ea: Le,
		aa: Re,
		U: ze,
		ka: We,
		T: Ge,
		ia: Je,
		ga: Qe,
		M: et,
		N: gt,
		s: vt,
		n: yt,
		ha: St,
		E: At,
		F: jt,
		la: Mt,
		ja: Nt,
		D: Vt,
		ma: Xe,
		Q: Ht,
		G: Ut,
		A: Wt,
		W: Gt,
		V: Kt,
		$: Xt,
		X: $t,
		Y: en,
		J: tn,
		ca: nn,
		S: an,
		K: cn,
		H: Hn,
		O: jn,
		I: Vn,
		l: Un,
		b: kn,
		c: Dn,
		f: An,
		j: Fn,
		C: Ln,
		r: zn,
		B: Bn,
		x: Wn,
		R: Kn,
		k: On,
		i: Cn,
		d: Tn,
		h: En,
		o: wn,
		y: Rn,
		z: Nn,
		q: In,
		fa: Pn,
		m: Mn,
		w: Gn,
		P: ln
	};
	function Cn(e, t) {
		var n = k();
		try {
			K(e)(t);
		} catch (e) {
			if (O(n), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function wn(e, t, n, r, i) {
		var a = k();
		try {
			K(e)(t, n, r, i);
		} catch (e) {
			if (O(a), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Tn(e, t, n) {
		var r = k();
		try {
			K(e)(t, n);
		} catch (e) {
			if (O(r), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function En(e, t, n, r) {
		var i = k();
		try {
			K(e)(t, n, r);
		} catch (e) {
			if (O(i), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Dn(e, t, n) {
		var r = k();
		try {
			return K(e)(t, n);
		} catch (e) {
			if (O(r), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function On(e) {
		var t = k();
		try {
			K(e)();
		} catch (e) {
			if (O(t), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function kn(e, t) {
		var n = k();
		try {
			return K(e)(t);
		} catch (e) {
			if (O(n), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function An(e, t, n, r) {
		var i = k();
		try {
			return K(e)(t, n, r);
		} catch (e) {
			if (O(i), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function jn(e, t, n, r, i, a) {
		var o = k();
		try {
			return K(e)(t, n, r, i, a);
		} catch (e) {
			if (O(o), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Mn(e, t, n, r, i, a, o, s, c, l, u) {
		var d = k();
		try {
			K(e)(t, n, r, i, a, o, s, c, l, u);
		} catch (e) {
			if (O(d), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Nn(e, t, n, r, i, a, o) {
		var s = k();
		try {
			K(e)(t, n, r, i, a, o);
		} catch (e) {
			if (O(s), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Pn(e, t, n, r, i, a, o, s, c) {
		var l = k();
		try {
			K(e)(t, n, r, i, a, o, s, c);
		} catch (e) {
			if (O(l), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Fn(e, t, n, r, i) {
		var a = k();
		try {
			return K(e)(t, n, r, i);
		} catch (e) {
			if (O(a), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function In(e, t, n, r, i, a, o, s) {
		var c = k();
		try {
			K(e)(t, n, r, i, a, o, s);
		} catch (e) {
			if (O(c), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Ln(e, t, n, r, i, a) {
		var o = k();
		try {
			return K(e)(t, n, r, i, a);
		} catch (e) {
			if (O(o), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Rn(e, t, n, r, i, a) {
		var o = k();
		try {
			K(e)(t, n, r, i, a);
		} catch (e) {
			if (O(o), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function zn(e, t, n, r, i, a, o) {
		var s = k();
		try {
			return K(e)(t, n, r, i, a, o);
		} catch (e) {
			if (O(s), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Bn(e, t, n, r, i, a, o, s) {
		var c = k();
		try {
			return K(e)(t, n, r, i, a, o, s);
		} catch (e) {
			if (O(c), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Vn(e, t, n, r) {
		var i = k();
		try {
			return K(e)(t, n, r);
		} catch (e) {
			if (O(i), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Hn(e, t, n, r) {
		var i = k();
		try {
			return K(e)(t, n, r);
		} catch (e) {
			if (O(i), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Un(e) {
		var t = k();
		try {
			return K(e)();
		} catch (e) {
			if (O(t), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Wn(e, t, n, r, i, a, o, s, c, l, u, d) {
		var f = k();
		try {
			return K(e)(t, n, r, i, a, o, s, c, l, u, d);
		} catch (e) {
			if (O(f), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Gn(e, t, n, r, i, a, o, s, c, l, u, d, f, p, m, h) {
		var g = k();
		try {
			K(e)(t, n, r, i, a, o, s, c, l, u, d, f, p, m, h);
		} catch (e) {
			if (O(g), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function Kn(e, t, n, r, i) {
		var a = k();
		try {
			return yn(e, t, n, r, i);
		} catch (e) {
			if (O(a), e !== e + 0) throw e;
			Z(1, 0);
		}
	}
	function qn() {
		ie();
		function e() {
			var e, t;
			i.calledRun = !0, !_ && (ae(), (e = ee) == null || e(i), (t = i.onRuntimeInitialized) == null || t.call(i), oe());
		}
		i.setStatus ? (i.setStatus("Running..."), setTimeout(() => {
			setTimeout(() => i.setStatus(""), 1), e();
		}, 1)) : e();
	}
	var $ = await de();
	return qn(), t = ne ? i : new Promise((e, t) => {
		ee = e, te = t;
	}), t;
}
//#endregion
//#region src/writer/index.ts
function x(e) {
	return p(b, e);
}
function se() {
	return re(b);
}
function ce(e) {
	return x({
		overrides: e,
		equalityFn: Object.is,
		fireImmediately: !0
	});
}
function le(e) {
	x({
		overrides: e,
		equalityFn: Object.is,
		fireImmediately: !1
	});
}
async function ue(e, t) {
	return h(b, e, t);
}
var de = "626d0633671ca3372faa058e565fae1ad985b8c0eb1b246f5fca5ff569e62a2f";
//#endregion
export { y as BARCODE_FORMATS, m as BARCODE_HRI_LABELS, v as BARCODE_META_FORMATS, d as BARCODE_SYMBOLOGIES, _ as CHARACTER_SETS, t as CREATABLE_BARCODE_FORMATS, oe as GS1_BARCODE_FORMATS, f as INDUSTRIAL_BARCODE_FORMATS, r as LINEAR_BARCODE_FORMATS, n as MATRIX_BARCODE_FORMATS, l as READABLE_BARCODE_FORMATS, ne as RETAIL_BARCODE_FORMATS, ie as ZXING_CPP_COMMIT, de as ZXING_WASM_SHA256, ee as ZXING_WASM_VERSION, e as barcodeFormats, g as characterSets, ae as defaultWriterOptions, te as encodeFormat, s as encodeFormats, c as formatToLabel, u as formatToSymbology, ce as getZXingModule, i as linearBarcodeFormats, a as matrixBarcodeFormats, x as prepareZXingModule, se as purgeZXingModule, le as setZXingModuleOverrides, o as symbologyToFormats, ue as writeBarcode };
