// zxing-wasm/dist/es/full/index.js
import { A as e, C as t, D as n, E as r, F as i, I as a, L as o, M as s, N as c, O as l, P as u, S as d, T as f, _ as p, a as m, b as h, c as g, d as _, f as ee, g as v, h as y, i as te, j as ne, k as re, l as ie, m as ae, n as b, o as oe, p as se, r as ce, s as le, t as ue, u as de, v as fe, w as x, x as S, y as C } from "../share.js";
//#region src/full/zxing_full.js
async function w(e = {}) {
	var t, n, r, i = e, a = !!globalThis.window, o = typeof Bun < "u", s = !!globalThis.WorkerGlobalScope;
	!((n = globalThis.process) == null || (n = n.versions) == null) && n.node && ((r = globalThis.process) == null || r.type);
	var c = "./this.program", l, u = "";
	function d(e) {
		return i.locateFile ? i.locateFile(e, u) : u + e;
	}
	var f, p;
	if (a || s || o) {
		try {
			u = new URL(".", l).href;
		} catch {}
		s && (p = (e) => {
			var t = new XMLHttpRequest();
			return t.open("GET", e, !1), t.responseType = "arraybuffer", t.send(null), new Uint8Array(t.response);
		}), f = async (e) => {
			var t = await fetch(e, { credentials: "same-origin" });
			if (t.ok) return t.arrayBuffer();
			throw Error(t.status + " : " + t.url);
		};
	}
	var m = console.log.bind(console), h = console.error.bind(console), g, _ = !1, ee, v, y = !1;
	function te() {
		var e = rr.buffer;
		C = new Int8Array(e), x = new Int16Array(e), i.HEAPU8 = O = new Uint8Array(e), E = new Uint16Array(e), S = new Int32Array(e), D = new Uint32Array(e), w = new Float32Array(e), T = new Float64Array(e);
	}
	function ne() {
		if (i.preRun) for (typeof i.preRun == "function" && (i.preRun = [i.preRun]); i.preRun.length;) ge(i.preRun.shift());
		k(he);
	}
	function re() {
		y = !0, Kr.Ha();
	}
	function ie() {
		if (i.postRun) for (typeof i.postRun == "function" && (i.postRun = [i.postRun]); i.postRun.length;) me(i.postRun.shift());
		k(pe);
	}
	function ae(e) {
		var t, n;
		(t = i.onAbort) == null || t.call(i, e), e = "Aborted(" + e + ")", h(e), _ = !0, e += ". Build with -sASSERTIONS for more info.";
		var r = new WebAssembly.RuntimeError(e);
		throw (n = v) == null || n(r), r;
	}
	var b;
	function oe() {
		return d("zxing_full.wasm");
	}
	function se(e) {
		if (e == b && g) return new Uint8Array(g);
		if (p) return p(e);
		throw "both async and sync fetching of the wasm failed";
	}
	async function ce(e) {
		if (!g) try {
			var t = await f(e);
			return new Uint8Array(t);
		} catch {}
		return se(e);
	}
	async function le(e, t) {
		try {
			var n = await ce(e);
			return await WebAssembly.instantiate(n, t);
		} catch (e) {
			h(`failed to asynchronously prepare wasm: ${e}`), ae(e);
		}
	}
	async function ue(e, t, n) {
		if (!e && WebAssembly.instantiateStreaming) try {
			var r = fetch(t, { credentials: "same-origin" });
			return await WebAssembly.instantiateStreaming(r, n);
		} catch (e) {
			h(`wasm streaming compile failed: ${e}`), h("falling back to ArrayBuffer instantiation");
		}
		return le(t, n);
	}
	function de() {
		return { a: or };
	}
	async function fe() {
		function e(e, t) {
			return Kr = e.exports, ar(Kr), te(), Kr;
		}
		function t(t) {
			return e(t.instance);
		}
		var n = de();
		return i.instantiateWasm ? new Promise((t, r) => {
			i.instantiateWasm(n, (n, r) => {
				t(e(n, r));
			});
		}) : (b != null || (b = oe()), t(await ue(g, b, n)));
	}
	var x, S, C, w, T, E, D, O, k = (e) => {
		for (; e.length > 0;) e.shift()(i);
	}, pe = [], me = (e) => pe.push(e), he = [], ge = (e) => he.push(e), A = (e) => Yn(e), j = () => Xn(), M = [], N = 0, _e = (e) => {
		var t = new ye(e);
		return t.get_caught() || (t.set_caught(!0), N--), t.set_rethrown(!1), M.push(t), qn(e);
	}, P = 0, ve = () => {
		$(0, 0);
		var e = M.pop();
		Zn(e.excPtr), P = 0;
	};
	class ye {
		constructor(e) {
			this.excPtr = e, this.ptr = e - 24;
		}
		set_type(e) {
			D[this.ptr + 4 >> 2] = e;
		}
		get_type() {
			return D[this.ptr + 4 >> 2];
		}
		set_destructor(e) {
			D[this.ptr + 8 >> 2] = e;
		}
		get_destructor() {
			return D[this.ptr + 8 >> 2];
		}
		set_caught(e) {
			e = +!!e, C[this.ptr + 12] = e;
		}
		get_caught() {
			return C[this.ptr + 12] != 0;
		}
		set_rethrown(e) {
			e = +!!e, C[this.ptr + 13] = e;
		}
		get_rethrown() {
			return C[this.ptr + 13] != 0;
		}
		init(e, t) {
			this.set_adjusted_ptr(0), this.set_type(e), this.set_destructor(t);
		}
		set_adjusted_ptr(e) {
			D[this.ptr + 16 >> 2] = e;
		}
		get_adjusted_ptr() {
			return D[this.ptr + 16 >> 2];
		}
	}
	var F = (e) => Jn(e), be = (e) => {
		var t = P;
		if (!t) return F(0), 0;
		var n = new ye(t);
		n.set_adjusted_ptr(t);
		var r = n.get_type();
		if (!r) return F(0), t;
		for (var i of e) {
			if (i === 0 || i === r) break;
			var a = n.ptr + 16;
			if ($n(i, r, a)) return F(i), t;
		}
		return F(r), t;
	}, xe = () => be([]), Se = (e) => be([e]), Ce = (e, t) => be([e, t]), we = () => {
		var e = M.pop();
		e || ae("no exception to throw");
		var t = e.excPtr;
		throw e.get_rethrown() || (M.push(e), e.set_rethrown(!0), e.set_caught(!1), N++), Qn(t), P = t, P;
	}, Te = (e, t, n) => {
		throw new ye(e).init(t, n), Qn(e), P = e, N++, P;
	}, Ee = () => N, De = (e) => {
		throw P || (P = e), P;
	}, Oe = globalThis.TextDecoder && new TextDecoder(), ke = (e, t, n, r) => {
		var i = t + n;
		if (r) return i;
		for (; e[t] && !(t >= i);) ++t;
		return t;
	}, Ae = function(e) {
		let t = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : 0, n = arguments.length > 2 ? arguments[2] : void 0, r = arguments.length > 3 ? arguments[3] : void 0;
		var i = ke(e, t, n, r);
		if (i - t > 16 && e.buffer && Oe) return Oe.decode(e.subarray(t, i));
		for (var a = ""; t < i;) {
			var o = e[t++];
			if (!(o & 128)) {
				a += String.fromCharCode(o);
				continue;
			}
			var s = e[t++] & 63;
			if ((o & 224) == 192) {
				a += String.fromCharCode((o & 31) << 6 | s);
				continue;
			}
			var c = e[t++] & 63;
			if (o = (o & 240) == 224 ? (o & 15) << 12 | s << 6 | c : (o & 7) << 18 | s << 12 | c << 6 | e[t++] & 63, o < 65536) a += String.fromCharCode(o);
			else {
				var l = o - 65536;
				a += String.fromCharCode(55296 | l >> 10, 56320 | l & 1023);
			}
		}
		return a;
	}, je = (e, t, n) => e ? Ae(O, e, t, n) : "", Me = {
		varargs: void 0,
		getStr(e) {
			return je(e);
		}
	};
	function Ne(e, t, n) {
		return Me.varargs = n, 0;
	}
	function Pe(e, t, n) {
		return Me.varargs = n, 0;
	}
	var Fe = (e, t, n) => {};
	function Ie(e, t, n, r) {
		Me.varargs = r;
	}
	var Le = (e, t) => {}, Re = () => ae(""), ze = {}, Be = (e) => {
		for (; e.length;) {
			var t = e.pop();
			e.pop()(t);
		}
	};
	function I(e) {
		return this.fromWireType(D[e >> 2]);
	}
	var L = {}, R = {}, Ve = {}, He = class extends Error {
		constructor(e) {
			super(e), this.name = "InternalError";
		}
	}, Ue = (e) => {
		throw new He(e);
	}, z = (e, t, n) => {
		e.forEach((e) => Ve[e] = t);
		function r(t) {
			var r = n(t);
			r.length !== e.length && Ue("Mismatched type converter count");
			for (var i = 0; i < e.length; ++i) U(e[i], r[i]);
		}
		var i = Array(t.length), a = [], o = 0;
		{
			let e = t;
			for (let t = 0; t < e.length; ++t) {
				let n = e[t];
				R.hasOwnProperty(n) ? i[t] = R[n] : (a.push(n), L.hasOwnProperty(n) || (L[n] = []), L[n].push(() => {
					i[t] = R[n], ++o, o === a.length && r(i);
				}));
			}
		}
		a.length === 0 && r(i);
	}, We = (e) => {
		var t = ze[e];
		delete ze[e];
		var n = t.rawConstructor, r = t.rawDestructor, i = t.fields, a = i.map((e) => e.getterReturnType).concat(i.map((e) => e.setterArgumentType));
		z([e], a, (e) => {
			var a = {};
			{
				let t = i;
				for (let n = 0; n < t.length; ++n) {
					let r = t[n], o = e[n], s = r.getter, c = r.getterContext, l = e[n + i.length], u = r.setter, d = r.setterContext;
					a[r.fieldName] = {
						read: (e) => o.fromWireType(s(c, e)),
						write: (e, t) => {
							var n = [];
							u(d, e, l.toWireType(n, t)), Be(n);
						},
						optional: o.optional
					};
				}
			}
			return [{
				name: t.name,
				fromWireType: (e) => {
					var t = {};
					for (var n in a) t[n] = a[n].read(e);
					return r(e), t;
				},
				toWireType: (e, t) => {
					for (var i in a) if (!(i in t) && !a[i].optional) throw TypeError(`Missing field: "${i}"`);
					var o = n();
					for (i in a) a[i].write(o, t[i]);
					return e !== null && e.push(r, o), o;
				},
				readValueFromPointer: I,
				destructorFunction: r
			}];
		});
	}, Ge = (e, t, n, r, i) => {}, B = (e) => {
		for (var t = "";;) {
			var n = O[e++];
			if (!n) return t;
			t += String.fromCharCode(n);
		}
	}, V = class extends Error {
		constructor(e) {
			super(e), this.name = "BindingError";
		}
	}, H = (e) => {
		throw new V(e);
	};
	function Ke(e, t) {
		let n = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {};
		var r = t.name;
		if (e || H(`type "${r}" must have a positive integer typeid pointer`), R.hasOwnProperty(e)) {
			if (n.ignoreDuplicateRegistrations) return;
			H(`Cannot register type '${r}' twice`);
		}
		if (R[e] = t, delete Ve[e], L.hasOwnProperty(e)) {
			var i = L[e];
			delete L[e], i.forEach((e) => e());
		}
	}
	function U(e, t) {
		return Ke(e, t, arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {});
	}
	var qe = (e, t, n, r) => {
		t = B(t), U(e, {
			name: t,
			fromWireType: function(e) {
				return !!e;
			},
			toWireType: function(e, t) {
				return t ? n : r;
			},
			readValueFromPointer: function(e) {
				return this.fromWireType(O[e]);
			},
			destructorFunction: null
		});
	}, Je = (e) => ({
		count: e.count,
		deleteScheduled: e.deleteScheduled,
		preservePointerOnDelete: e.preservePointerOnDelete,
		ptr: e.ptr,
		ptrType: e.ptrType,
		smartPtr: e.smartPtr,
		smartPtrType: e.smartPtrType
	}), Ye = (e) => {
		function t(e) {
			return e.$$.ptrType.registeredClass.name;
		}
		H(t(e) + " instance already deleted");
	}, Xe = !1, Ze = (e) => {}, Qe = (e) => {
		e.smartPtr ? e.smartPtrType.rawDestructor(e.smartPtr) : e.ptrType.registeredClass.rawDestructor(e.ptr);
	}, $e = (e) => {
		--e.count.value, e.count.value === 0 && Qe(e);
	}, W = (e) => globalThis.FinalizationRegistry ? (Xe = new FinalizationRegistry((e) => {
		$e(e.$$);
	}), W = (e) => {
		var t = e.$$;
		if (t.smartPtr) {
			var n = { $$: t };
			Xe.register(e, n, e);
		}
		return e;
	}, Ze = (e) => Xe.unregister(e), W(e)) : (W = (e) => e, e), et = [], tt = () => {
		for (; et.length;) {
			var e = et.pop();
			e.$$.deleteScheduled = !1, e.delete();
		}
	}, nt, rt = () => {
		let e = it.prototype;
		Object.assign(e, {
			isAliasOf(e) {
				if (!(this instanceof it) || !(e instanceof it)) return !1;
				var t = this.$$.ptrType.registeredClass, n = this.$$.ptr;
				e.$$ = e.$$;
				for (var r = e.$$.ptrType.registeredClass, i = e.$$.ptr; t.baseClass;) n = t.upcast(n), t = t.baseClass;
				for (; r.baseClass;) i = r.upcast(i), r = r.baseClass;
				return t === r && n === i;
			},
			clone() {
				if (this.$$.ptr || Ye(this), this.$$.preservePointerOnDelete) return this.$$.count.value += 1, this;
				var e = W(Object.create(Object.getPrototypeOf(this), { $$: { value: Je(this.$$) } }));
				return e.$$.count.value += 1, e.$$.deleteScheduled = !1, e;
			},
			delete() {
				this.$$.ptr || Ye(this), this.$$.deleteScheduled && !this.$$.preservePointerOnDelete && H("Object already scheduled for deletion"), Ze(this), $e(this.$$), this.$$.preservePointerOnDelete || (this.$$.smartPtr = void 0, this.$$.ptr = void 0);
			},
			isDeleted() {
				return !this.$$.ptr;
			},
			deleteLater() {
				return this.$$.ptr || Ye(this), this.$$.deleteScheduled && !this.$$.preservePointerOnDelete && H("Object already scheduled for deletion"), et.push(this), et.length === 1 && nt && nt(tt), this.$$.deleteScheduled = !0, this;
			}
		});
		let t = Symbol.dispose;
		t && (e[t] = e.delete);
	};
	function it() {}
	var at = (e, t) => Object.defineProperty(t, "name", { value: e }), ot = {}, st = (e, t, n) => {
		if (e[t].overloadTable === void 0) {
			var r = e[t];
			e[t] = function() {
				var r = [...arguments];
				return e[t].overloadTable.hasOwnProperty(r.length) || H(`Function '${n}' called with an invalid number of arguments (${r.length}) - expects one of (${e[t].overloadTable})!`), e[t].overloadTable[r.length].apply(this, r);
			}, e[t].overloadTable = [], e[t].overloadTable[r.argCount] = r;
		}
	}, ct = (e, t, n) => {
		i.hasOwnProperty(e) ? ((n === void 0 || i[e].overloadTable !== void 0 && i[e].overloadTable[n] !== void 0) && H(`Cannot register public name '${e}' twice`), st(i, e, e), i[e].overloadTable.hasOwnProperty(n) && H(`Cannot register multiple overloads of a function with the same number of arguments (${n})!`), i[e].overloadTable[n] = t) : (i[e] = t, i[e].argCount = n);
	}, lt = 48, ut = 57, dt = (e) => {
		e = e.replace(/[^a-zA-Z0-9_]/g, "$");
		var t = e.charCodeAt(0);
		return t >= lt && t <= ut ? `_${e}` : e;
	};
	function ft(e, t, n, r, i, a, o, s) {
		this.name = e, this.constructor = t, this.instancePrototype = n, this.rawDestructor = r, this.baseClass = i, this.getActualType = a, this.upcast = o, this.downcast = s, this.pureVirtualFunctions = [];
	}
	var pt = (e, t, n) => {
		for (; t !== n;) t.upcast || H(`Expected null or instance of ${n.name}, got an instance of ${t.name}`), e = t.upcast(e), t = t.baseClass;
		return e;
	}, mt = (e) => {
		if (e === null) return "null";
		var t = typeof e;
		return t === "object" || t === "array" || t === "function" ? e.toString() : "" + e;
	};
	function ht(e, t) {
		if (t === null) return this.isReference && H(`null is not a valid ${this.name}`), 0;
		t.$$ || H(`Cannot pass "${mt(t)}" as a ${this.name}`), t.$$.ptr || H(`Cannot pass deleted object as a pointer of type ${this.name}`);
		var n = t.$$.ptrType.registeredClass;
		return pt(t.$$.ptr, n, this.registeredClass);
	}
	function gt(e, t) {
		var n;
		if (t === null) return this.isReference && H(`null is not a valid ${this.name}`), this.isSmartPointer ? (n = this.rawConstructor(), e !== null && e.push(this.rawDestructor, n), n) : 0;
		(!t || !t.$$) && H(`Cannot pass "${mt(t)}" as a ${this.name}`), t.$$.ptr || H(`Cannot pass deleted object as a pointer of type ${this.name}`), !this.isConst && t.$$.ptrType.isConst && H(`Cannot convert argument of type ${t.$$.smartPtrType ? t.$$.smartPtrType.name : t.$$.ptrType.name} to parameter type ${this.name}`);
		var r = t.$$.ptrType.registeredClass;
		if (n = pt(t.$$.ptr, r, this.registeredClass), this.isSmartPointer) switch (t.$$.smartPtr === void 0 && H("Passing raw pointer to smart pointer is illegal"), this.sharingPolicy) {
			case 0:
				t.$$.smartPtrType === this ? n = t.$$.smartPtr : H(`Cannot convert argument of type ${t.$$.smartPtrType ? t.$$.smartPtrType.name : t.$$.ptrType.name} to parameter type ${this.name}`);
				break;
			case 1:
				n = t.$$.smartPtr;
				break;
			case 2:
				if (t.$$.smartPtrType === this) n = t.$$.smartPtr;
				else {
					var i = t.clone();
					n = this.rawShare(n, Y.toHandle(() => i.delete())), e !== null && e.push(this.rawDestructor, n);
				}
				break;
			default: H("Unsupported sharing policy");
		}
		return n;
	}
	function _t(e, t) {
		if (t === null) return this.isReference && H(`null is not a valid ${this.name}`), 0;
		t.$$ || H(`Cannot pass "${mt(t)}" as a ${this.name}`), t.$$.ptr || H(`Cannot pass deleted object as a pointer of type ${this.name}`), t.$$.ptrType.isConst && H(`Cannot convert argument of type ${t.$$.ptrType.name} to parameter type ${this.name}`);
		var n = t.$$.ptrType.registeredClass;
		return pt(t.$$.ptr, n, this.registeredClass);
	}
	var vt = (e, t, n) => {
		if (t === n) return e;
		if (n.baseClass === void 0) return null;
		var r = vt(e, t, n.baseClass);
		return r === null ? null : n.downcast(r);
	}, yt = {}, bt = (e, t) => {
		for (t === void 0 && H("ptr should not be undefined"); e.baseClass;) t = e.upcast(t), e = e.baseClass;
		return t;
	}, xt = (e, t) => (t = bt(e, t), yt[t]), St = (e, t) => ((!t.ptrType || !t.ptr) && Ue("makeClassHandle requires ptr and ptrType"), !!t.smartPtrType != !!t.smartPtr && Ue("Both smartPtrType and smartPtr must be specified"), t.count = { value: 1 }, W(Object.create(e, { $$: {
		value: t,
		writable: !0
	} })));
	function Ct(e) {
		var t = this.getPointee(e);
		if (!t) return this.destructor(e), null;
		var n = xt(this.registeredClass, t);
		if (n !== void 0) {
			if (n.$$.count.value === 0) return n.$$.ptr = t, n.$$.smartPtr = e, n.clone();
			var r = n.clone();
			return this.destructor(e), r;
		}
		function i() {
			return this.isSmartPointer ? St(this.registeredClass.instancePrototype, {
				ptrType: this.pointeeType,
				ptr: t,
				smartPtrType: this,
				smartPtr: e
			}) : St(this.registeredClass.instancePrototype, {
				ptrType: this,
				ptr: e
			});
		}
		var a = ot[this.registeredClass.getActualType(t)];
		if (!a) return i.call(this);
		var o = this.isConst ? a.constPointerType : a.pointerType, s = vt(t, this.registeredClass, o.registeredClass);
		return s === null ? i.call(this) : this.isSmartPointer ? St(o.registeredClass.instancePrototype, {
			ptrType: o,
			ptr: s,
			smartPtrType: this,
			smartPtr: e
		}) : St(o.registeredClass.instancePrototype, {
			ptrType: o,
			ptr: s
		});
	}
	var wt = () => {
		Object.assign(Tt.prototype, {
			getPointee(e) {
				return this.rawGetPointee && (e = this.rawGetPointee(e)), e;
			},
			destructor(e) {
				var t;
				(t = this.rawDestructor) == null || t.call(this, e);
			},
			readValueFromPointer: I,
			fromWireType: Ct
		});
	};
	function Tt(e, t, n, r, i, a, o, s, c, l, u) {
		this.name = e, this.registeredClass = t, this.isReference = n, this.isConst = r, this.isSmartPointer = i, this.pointeeType = a, this.sharingPolicy = o, this.rawGetPointee = s, this.rawConstructor = c, this.rawShare = l, this.rawDestructor = u, !i && t.baseClass === void 0 ? r ? (this.toWireType = ht, this.destructorFunction = null) : (this.toWireType = _t, this.destructorFunction = null) : this.toWireType = gt;
	}
	var Et = (e, t, n) => {
		i.hasOwnProperty(e) || Ue("Replacing nonexistent public symbol"), i[e].overloadTable !== void 0 && n !== void 0 ? i[e].overloadTable[n] = t : (i[e] = t, i[e].argCount = n);
	}, G = {}, Dt = (e, t, n) => {
		e = e.replace(/p/g, "i");
		var r = G[e];
		return r(t, ...n);
	}, Ot = [], K = (e) => {
		var t = Ot[e];
		return t || (Ot[e] = t = ir.get(e)), t;
	}, kt = function(e, t) {
		let n = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : [];
		if (arguments.length > 3 && arguments[3] !== void 0 && arguments[3], e.includes("j")) return Dt(e, t, n);
		var r = K(t)(...n);
		function i(e) {
			return e;
		}
		return i(r);
	}, At = function(e, t) {
		let n = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : !1;
		return function() {
			return kt(e, t, [...arguments], n);
		};
	}, q = function(e, t) {
		arguments.length > 2 && arguments[2] !== void 0 && arguments[2], e = B(e);
		function n() {
			return e.includes("j") ? At(e, t) : K(t);
		}
		var r = n();
		return typeof r != "function" && H(`unknown function pointer with signature ${e}: ${t}`), r;
	};
	class jt extends Error {}
	var Mt = (e) => {
		var t = Gn(e), n = B(t);
		return Q(t), n;
	}, Nt = (e, t) => {
		var n = [], r = {};
		function i(e) {
			if (!r[e] && !R[e]) {
				if (Ve[e]) {
					Ve[e].forEach(i);
					return;
				}
				n.push(e), r[e] = !0;
			}
		}
		throw t.forEach(i), new jt(`${e}: ` + n.map(Mt).join([", "]));
	}, Pt = (e, t, n, r, i, a, o, s, c, l, u, d, f) => {
		u = B(u), a = q(i, a), s && (s = q(o, s)), l && (l = q(c, l)), f = q(d, f);
		var p = dt(u);
		ct(p, function() {
			Nt(`Cannot construct ${u} due to unbound types`, [r]);
		}), z([
			e,
			t,
			n
		], r ? [r] : [], (t) => {
			t = t[0];
			var n, i;
			r ? (n = t.registeredClass, i = n.instancePrototype) : i = it.prototype;
			var o = at(u, function() {
				if (Object.getPrototypeOf(this) !== c) throw new V(`Use 'new' to construct ${u}`);
				if (d.constructor_body === void 0) throw new V(`${u} has no accessible constructor`);
				var e = [...arguments], t = d.constructor_body[e.length];
				if (t === void 0) throw new V(`Tried to invoke ctor of ${u} with invalid number of parameters (${e.length}) - expected (${Object.keys(d.constructor_body).toString()}) parameters instead!`);
				return t.apply(this, e);
			}), c = Object.create(i, { constructor: { value: o } });
			o.prototype = c;
			var d = new ft(u, o, c, f, n, a, s, l);
			if (d.baseClass) {
				var m;
				(m = d.baseClass).__derivedClasses != null || (m.__derivedClasses = []), d.baseClass.__derivedClasses.push(d);
			}
			var h = new Tt(u, d, !0, !1, !1), g = new Tt(u + "*", d, !1, !1, !1), _ = new Tt(u + " const*", d, !1, !0, !1);
			return ot[e] = {
				pointerType: g,
				constPointerType: _
			}, Et(p, o), [
				h,
				g,
				_
			];
		});
	}, Ft = (e, t) => {
		for (var n = [], r = 0; r < e; r++) n.push(D[t + r * 4 >> 2]);
		return n;
	};
	function It(e) {
		for (var t = 1; t < e.length; ++t) if (e[t] !== null && e[t].destructorFunction === void 0) return !0;
		return !1;
	}
	function Lt(e, t, n, r, i, a) {
		var o = t.length;
		o < 2 && H("argTypes array size mismatch! Must at least get return value and 'this' types!");
		var s = t[1] !== null && n !== null, c = It(t), l = !t[0].isVoid, u = o - 2, d = Array(u), f = [], p = [];
		return at(e, function() {
			p.length = 0;
			var e;
			f.length = s ? 2 : 1, f[0] = i, s && (e = t[1].toWireType(p, this), f[1] = e);
			for (var n = 0; n < u; ++n) d[n] = t[n + 2].toWireType(p, n < 0 || arguments.length <= n ? void 0 : arguments[n]), f.push(d[n]);
			var a = r(...f);
			function o(n) {
				if (c) Be(p);
				else for (var r = s ? 1 : 2; r < t.length; r++) {
					var i = r === 1 ? e : d[r - 2];
					t[r].destructorFunction !== null && t[r].destructorFunction(i);
				}
				if (l) return t[0].fromWireType(n);
			}
			return o(a);
		});
	}
	var Rt = (e, t, n, r, i, a) => {
		var o = Ft(t, n);
		i = q(r, i), z([], [e], (e) => {
			e = e[0];
			var n = `constructor ${e.name}`;
			if (e.registeredClass.constructor_body === void 0 && (e.registeredClass.constructor_body = []), e.registeredClass.constructor_body[t - 1] !== void 0) throw new V(`Cannot register multiple constructors with identical number of parameters (${t - 1}) for class '${e.name}'! Overload resolution is currently only performed using the parameter count, not actual type info!`);
			return e.registeredClass.constructor_body[t - 1] = () => {
				Nt(`Cannot construct ${e.name} due to unbound types`, o);
			}, z([], o, (r) => (r.splice(1, 0, null), e.registeredClass.constructor_body[t - 1] = Lt(n, r, null, i, a), [])), [];
		});
	}, zt = (e) => {
		e = e.trim();
		let t = e.indexOf("(");
		return t === -1 ? e : e.slice(0, t);
	}, Bt = (e, t, n, r, i, a, o, s, c, l) => {
		var u = Ft(n, r);
		t = B(t), t = zt(t), a = q(i, a, c), z([], [e], (e) => {
			e = e[0];
			var r = `${e.name}.${t}`;
			t.startsWith("@@") && (t = Symbol[t.substring(2)]), s && e.registeredClass.pureVirtualFunctions.push(t);
			function i() {
				Nt(`Cannot call ${r} due to unbound types`, u);
			}
			var l = e.registeredClass.instancePrototype, d = l[t];
			return d === void 0 || d.overloadTable === void 0 && d.className !== e.name && d.argCount === n - 2 ? (i.argCount = n - 2, i.className = e.name, l[t] = i) : (st(l, t, r), l[t].overloadTable[n - 2] = i), z([], u, (i) => {
				var s = Lt(r, i, e, a, o, c);
				return l[t].overloadTable === void 0 ? (s.argCount = n - 2, l[t] = s) : l[t].overloadTable[n - 2] = s, [];
			}), [];
		});
	}, Vt = [], J = [
		0,
		1,
		,
		1,
		null,
		1,
		!0,
		1,
		!1,
		1
	], Ht = (e) => {
		e > 9 && --J[e + 1] === 0 && (J[e] = void 0, Vt.push(e));
	}, Y = {
		toValue: (e) => (e || H(`Cannot use deleted val. handle = ${e}`), J[e]),
		toHandle: (e) => {
			switch (e) {
				case void 0: return 2;
				case null: return 4;
				case !0: return 6;
				case !1: return 8;
				default: {
					let t = Vt.pop() || J.length;
					return J[t] = e, J[t + 1] = 1, t;
				}
			}
		}
	}, Ut = {
		name: "emscripten::val",
		fromWireType: (e) => {
			var t = Y.toValue(e);
			return Ht(e), t;
		},
		toWireType: (e, t) => Y.toHandle(t),
		readValueFromPointer: I,
		destructorFunction: null
	}, Wt = (e) => U(e, Ut), Gt = (e, t) => {
		switch (t) {
			case 4: return function(e) {
				return this.fromWireType(w[e >> 2]);
			};
			case 8: return function(e) {
				return this.fromWireType(T[e >> 3]);
			};
			default: throw TypeError(`invalid float width (${t}): ${e}`);
		}
	}, Kt = (e, t, n) => {
		t = B(t), U(e, {
			name: t,
			fromWireType: (e) => e,
			toWireType: (e, t) => t,
			readValueFromPointer: Gt(t, n),
			destructorFunction: null
		});
	}, qt = (e, t, n, r, i, a, o, s) => {
		var c = Ft(t, n);
		e = B(e), e = zt(e), i = q(r, i, o), ct(e, function() {
			Nt(`Cannot call ${e} due to unbound types`, c);
		}, t - 1), z([], c, (n) => {
			var r = [n[0], null].concat(n.slice(1));
			return Et(e, Lt(e, r, null, i, a, o), t - 1), [];
		});
	}, Jt = (e, t, n) => {
		switch (t) {
			case 1: return n ? (e) => C[e] : (e) => O[e];
			case 2: return n ? (e) => x[e >> 1] : (e) => E[e >> 1];
			case 4: return n ? (e) => S[e >> 2] : (e) => D[e >> 2];
			default: throw TypeError(`invalid integer width (${t}): ${e}`);
		}
	}, Yt = (e, t, n, r, i) => {
		t = B(t);
		let a = r === 0, o = (e) => e;
		if (a) {
			var s = 32 - 8 * n;
			o = (e) => e << s >>> s, i = o(i);
		}
		U(e, {
			name: t,
			fromWireType: o,
			toWireType: (e, t) => t,
			readValueFromPointer: Jt(t, n, r !== 0),
			destructorFunction: null
		});
	}, Xt = (e, t, n) => {
		let r = (e, t) => {
			let n = 0;
			return {
				next() {
					if (n >= e) return { done: !0 };
					let r = n;
					return n++, {
						value: t(r),
						done: !1
					};
				},
				[Symbol.iterator]() {
					return this;
				}
			};
		};
		e[Symbol.iterator] || (e[Symbol.iterator] = function() {
			return r(this[t](), (e) => this[n](e));
		});
	}, Zt = (e, t, n, r) => {
		n = B(n), r = B(r), z([], [e, t], (e) => {
			let t = e[0];
			return Xt(t.registeredClass.instancePrototype, n, r), [];
		});
	}, Qt = (e, t, n) => {
		var r = [
			Int8Array,
			Uint8Array,
			Int16Array,
			Uint16Array,
			Int32Array,
			Uint32Array,
			Float32Array,
			Float64Array
		][t];
		function i(e) {
			var t = D[e >> 2], n = D[e + 4 >> 2];
			return new r(C.buffer, n, t);
		}
		n = B(n), U(e, {
			name: n,
			fromWireType: i,
			readValueFromPointer: i
		}, { ignoreDuplicateRegistrations: !0 });
	}, $t = Object.assign({ optional: !0 }, Ut), en = (e, t) => {
		U(e, $t);
	}, tn = (e, t, n, r) => {
		if (!(r > 0)) return 0;
		for (var i = n, a = n + r - 1, o = 0; o < e.length; ++o) {
			var s = e.codePointAt(o);
			if (s <= 127) {
				if (n >= a) break;
				t[n++] = s;
			} else if (s <= 2047) {
				if (n + 1 >= a) break;
				t[n++] = 192 | s >> 6, t[n++] = 128 | s & 63;
			} else if (s <= 65535) {
				if (n + 2 >= a) break;
				t[n++] = 224 | s >> 12, t[n++] = 128 | s >> 6 & 63, t[n++] = 128 | s & 63;
			} else {
				if (n + 3 >= a) break;
				t[n++] = 240 | s >> 18, t[n++] = 128 | s >> 12 & 63, t[n++] = 128 | s >> 6 & 63, t[n++] = 128 | s & 63, o++;
			}
		}
		return t[n] = 0, n - i;
	}, X = (e, t, n) => tn(e, O, t, n), nn = (e) => {
		for (var t = 0, n = 0; n < e.length; ++n) {
			var r = e.charCodeAt(n);
			r <= 127 ? t++ : r <= 2047 ? t += 2 : r >= 55296 && r <= 57343 ? (t += 4, ++n) : t += 3;
		}
		return t;
	}, rn = (e, t) => {
		t = B(t);
		var n = !0;
		U(e, {
			name: t,
			fromWireType(e) {
				var t = D[e >> 2], r = e + 4, i;
				if (n) i = je(r, t, !0);
				else {
					i = "";
					for (var a = 0; a < t; ++a) i += String.fromCharCode(O[r + a]);
				}
				return Q(e), i;
			},
			toWireType(e, t) {
				t instanceof ArrayBuffer && (t = new Uint8Array(t));
				var r, i = typeof t == "string";
				i || ArrayBuffer.isView(t) && t.BYTES_PER_ELEMENT == 1 || H("Cannot pass non-string to std::string"), r = n && i ? nn(t) : t.length;
				var a = Kn(4 + r + 1), o = a + 4;
				if (D[a >> 2] = r, i) if (n) X(t, o, r + 1);
				else for (var s = 0; s < r; ++s) {
					var c = t.charCodeAt(s);
					c > 255 && (Q(a), H("String has UTF-16 code units that do not fit in 8 bits")), O[o + s] = c;
				}
				else O.set(t, o);
				return e !== null && e.push(Q, a), a;
			},
			readValueFromPointer: I,
			destructorFunction(e) {
				Q(e);
			}
		});
	}, an = globalThis.TextDecoder ? new TextDecoder("utf-16le") : void 0, on = (e, t, n) => {
		var r = e >> 1, i = ke(E, r, t / 2, n);
		if (i - r > 16 && an) return an.decode(E.subarray(r, i));
		for (var a = "", o = r; o < i; ++o) {
			var s = E[o];
			a += String.fromCharCode(s);
		}
		return a;
	}, sn = (e, t, n) => {
		if (n != null || (n = 2147483647), n < 2) return 0;
		n -= 2;
		for (var r = t, i = n < e.length * 2 ? n / 2 : e.length, a = 0; a < i; ++a) {
			var o = e.charCodeAt(a);
			x[t >> 1] = o, t += 2;
		}
		return x[t >> 1] = 0, t - r;
	}, cn = (e) => e.length * 2, ln = (e, t, n) => {
		for (var r = "", i = e >> 2, a = 0; !(a >= t / 4); a++) {
			var o = D[i + a];
			if (!o && !n) break;
			r += String.fromCodePoint(o);
		}
		return r;
	}, un = (e, t, n) => {
		if (n != null || (n = 2147483647), n < 4) return 0;
		for (var r = t, i = r + n - 4, a = 0; a < e.length; ++a) {
			var o = e.codePointAt(a);
			if (o > 65535 && a++, S[t >> 2] = o, t += 4, t + 4 > i) break;
		}
		return S[t >> 2] = 0, t - r;
	}, dn = (e) => {
		for (var t = 0, n = 0; n < e.length; ++n) e.codePointAt(n) > 65535 && n++, t += 4;
		return t;
	}, fn = (e, t, n) => {
		n = B(n);
		var r, i, a;
		t === 2 ? (r = on, i = sn, a = cn) : (r = ln, i = un, a = dn), U(e, {
			name: n,
			fromWireType: (e) => {
				var n = D[e >> 2], i = r(e + 4, n * t, !0);
				return Q(e), i;
			},
			toWireType: (e, r) => {
				typeof r != "string" && H(`Cannot pass non-string to C++ string type ${n}`);
				var o = a(r), s = Kn(4 + o + t);
				return D[s >> 2] = o / t, i(r, s + 4, o + t), e !== null && e.push(Q, s), s;
			},
			readValueFromPointer: I,
			destructorFunction(e) {
				Q(e);
			}
		});
	}, pn = (e, t, n, r, i, a) => {
		ze[e] = {
			name: B(t),
			rawConstructor: q(n, r),
			rawDestructor: q(i, a),
			fields: []
		};
	}, mn = (e, t, n, r, i, a, o, s, c, l) => {
		ze[e].fields.push({
			fieldName: B(t),
			getterReturnType: n,
			getter: q(r, i),
			getterContext: a,
			setterArgumentType: o,
			setter: q(s, c),
			setterContext: l
		});
	}, hn = (e, t) => {
		t = B(t), U(e, {
			isVoid: !0,
			name: t,
			fromWireType: () => void 0,
			toWireType: (e, t) => void 0
		});
	}, gn = [], _n = (e) => {
		var t = gn.length;
		return gn.push(e), t;
	}, vn = (e, t) => {
		var n = R[e];
		return n === void 0 && H(`${t} has unknown type ${Mt(e)}`), n;
	}, yn = (e, t) => {
		for (var n = Array(e), r = 0; r < e; ++r) n[r] = vn(D[t + r * 4 >> 2], `parameter ${r}`);
		return n;
	}, bn = (e, t, n) => {
		var r = [], i = e(r, n);
		return r.length && (D[t >> 2] = Y.toHandle(r)), i;
	}, xn = {}, Sn = (e) => {
		var t = xn[e];
		return t === void 0 ? B(e) : t;
	}, Cn = (e, t, n) => {
		var r = 8, [i, ...a] = yn(e, t), o = i.toWireType.bind(i), s = a.map((e) => e.readValueFromPointer.bind(e));
		e--;
		var c = Array(e);
		return _n(at(`methodCaller<(${a.map((e) => e.name)}) => ${i.name}>`, (t, i, a, l) => {
			for (var u = 0, d = 0; d < e; ++d) c[d] = s[d](l + u), u += r;
			var f;
			switch (n) {
				case 0:
					f = Y.toValue(t).apply(null, c);
					break;
				case 2:
					f = Reflect.construct(Y.toValue(t), c);
					break;
				case 3:
					f = c[0];
					break;
				case 1:
					f = Y.toValue(t)[Sn(i)](...c);
					break;
			}
			return bn(o, a, f);
		}));
	}, wn = (e) => e ? (e = Sn(e), Y.toHandle(globalThis[e])) : Y.toHandle(globalThis), Tn = (e) => {
		e > 9 && (J[e + 1] += 1);
	}, En = (e, t, n, r, i) => gn[e](t, n, r, i), Dn = (e) => {
		Be(Y.toValue(e)), Ht(e);
	}, On = (e, t, n, r) => {
		var i = (/* @__PURE__ */ new Date()).getFullYear(), a = new Date(i, 0, 1), o = new Date(i, 6, 1), s = a.getTimezoneOffset(), c = o.getTimezoneOffset(), l = Math.max(s, c);
		D[e >> 2] = l * 60, S[t >> 2] = Number(s != c);
		var u = (e) => {
			var t = e >= 0 ? "-" : "+", n = Math.abs(e);
			return `UTC${t}${String(Math.floor(n / 60)).padStart(2, "0")}${String(n % 60).padStart(2, "0")}`;
		}, d = u(s), f = u(c);
		c < s ? (X(d, n, 17), X(f, r, 17)) : (X(d, r, 17), X(f, n, 17));
	}, kn = () => 2147483648, An = (e, t) => Math.ceil(e / t) * t, jn = (e) => {
		var t = (e - rr.buffer.byteLength + 65535) / 65536 | 0;
		try {
			return rr.grow(t), te(), 1;
		} catch {}
	}, Mn = (e) => {
		var t = O.length;
		e >>>= 0;
		var n = kn();
		if (e > n) return !1;
		for (var r = 1; r <= 4; r *= 2) {
			var i = t * (1 + .2 / r);
			if (i = Math.min(i, e + 100663296), jn(Math.min(n, An(Math.max(e, i), 65536)))) return !0;
		}
		return !1;
	}, Nn = {}, Pn = () => c || "./this.program", Z = () => {
		if (!Z.strings) {
			var e, t, n = {
				USER: "web_user",
				LOGNAME: "web_user",
				PATH: "/",
				PWD: "/",
				HOME: "/home/web_user",
				LANG: ((e = (t = globalThis.navigator) == null ? void 0 : t.language) == null ? "C" : e).replace("-", "_") + ".UTF-8",
				_: Pn()
			};
			for (var r in Nn) Nn[r] === void 0 ? delete n[r] : n[r] = Nn[r];
			var i = [];
			for (var r in n) i.push(`${r}=${n[r]}`);
			Z.strings = i;
		}
		return Z.strings;
	}, Fn = (e, t) => {
		var n = 0, r = 0;
		for (var i of Z()) {
			var a = t + n;
			D[e + r >> 2] = a, n += X(i, a, Infinity) + 1, r += 4;
		}
		return 0;
	}, In = (e, t) => {
		var n = Z();
		D[e >> 2] = n.length;
		var r = 0;
		for (var i of n) r += nn(i) + 1;
		return D[t >> 2] = r, 0;
	}, Ln = (e) => 52, Rn = (e, t, n, r) => 52, zn = (e, t) => t + 2097152 >>> 0 < 4194305 - !!e ? (e >>> 0) + t * 4294967296 : NaN;
	function Bn(e, t, n, r, i) {
		return zn(t, n), 70;
	}
	var Vn = [
		null,
		[],
		[]
	], Hn = (e, t) => {
		var n = Vn[e];
		t === 0 || t === 10 ? ((e === 1 ? m : h)(Ae(n)), n.length = 0) : n.push(t);
	}, Un = (e, t, n, r) => {
		for (var i = 0, a = 0; a < n; a++) {
			var o = D[t >> 2], s = D[t + 4 >> 2];
			t += 8;
			for (var c = 0; c < s; c++) Hn(e, O[o + c]);
			i += s;
		}
		return D[r >> 2] = i, 0;
	}, Wn = (e) => e;
	if (rt(), wt(), i.noExitRuntime && i.noExitRuntime, i.print && (m = i.print), i.printErr && (h = i.printErr), i.wasmBinary && (g = i.wasmBinary), i.arguments && i.arguments, i.thisProgram && (c = i.thisProgram), i.preInit) for (typeof i.preInit == "function" && (i.preInit = [i.preInit]); i.preInit.length > 0;) i.preInit.shift()();
	var Gn, Q, Kn, qn, $, Jn, Yn, Xn, Zn, Qn, $n, er, tr, nr, rr, ir;
	function ar(e) {
		Gn = e.Ia, Q = i._free = e.Ja, Kn = i._malloc = e.La, qn = e.Ma, $ = e.Na, Jn = e.Oa, Yn = e.Pa, Xn = e.Qa, Zn = e.Ra, Qn = e.Sa, $n = e.Ta, G.viijii = e.Ua, er = G.viijjijjjjjj = e.Va, tr = G.iiijj = e.Wa, G.jiji = e.Xa, nr = G.jiiii = e.Ya, G.iiiiij = e.Za, G.iiiiijj = e._a, G.iiiiiijj = e.$a, rr = e.Ga, ir = e.Ka;
	}
	var or = {
		s: _e,
		x: ve,
		a: xe,
		i: Se,
		m: Ce,
		S: we,
		p: Te,
		ia: Ee,
		d: De,
		Y: Ne,
		na: Pe,
		la: Fe,
		oa: Ie,
		ka: Le,
		ea: Re,
		Da: We,
		da: Ge,
		xa: qe,
		Ba: Pt,
		Aa: Rt,
		J: Bt,
		va: Wt,
		_: Kt,
		Q: qt,
		A: Yt,
		za: Zt,
		u: Qt,
		Ca: en,
		wa: rn,
		T: fn,
		C: pn,
		Ea: mn,
		ya: hn,
		P: Cn,
		Fa: Ht,
		H: wn,
		U: Tn,
		O: En,
		ra: Dn,
		fa: On,
		ja: Mn,
		ga: Fn,
		ha: In,
		W: Ln,
		ma: Rn,
		aa: Bn,
		X: Un,
		ta: Or,
		N: jr,
		D: Lr,
		G: mr,
		V: zr,
		q: Dr,
		b: lr,
		I: Ar,
		qa: Fr,
		c: dr,
		pa: Ir,
		g: pr,
		j: br,
		r: Sr,
		R: kr,
		t: wr,
		E: Tr,
		B: Er,
		L: Br,
		ba: Ur,
		$: Wr,
		f: hr,
		l: sr,
		e: ur,
		Z: Mr,
		h: fr,
		M: Rr,
		k: cr,
		sa: Nr,
		o: Cr,
		y: _r,
		v: xr,
		F: yr,
		w: Pr,
		n: gr,
		K: Vr,
		ua: vr,
		ca: Hr,
		z: Wn
	};
	function sr(e, t) {
		var n = j();
		try {
			K(e)(t);
		} catch (e) {
			if (A(n), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function cr(e, t, n, r, i) {
		var a = j();
		try {
			K(e)(t, n, r, i);
		} catch (e) {
			if (A(a), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function lr(e, t) {
		var n = j();
		try {
			return K(e)(t);
		} catch (e) {
			if (A(n), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function ur(e, t, n) {
		var r = j();
		try {
			K(e)(t, n);
		} catch (e) {
			if (A(r), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function dr(e, t, n) {
		var r = j();
		try {
			return K(e)(t, n);
		} catch (e) {
			if (A(r), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function fr(e, t, n, r) {
		var i = j();
		try {
			K(e)(t, n, r);
		} catch (e) {
			if (A(i), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function pr(e, t, n, r) {
		var i = j();
		try {
			return K(e)(t, n, r);
		} catch (e) {
			if (A(i), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function mr(e, t, n, r, i, a) {
		var o = j();
		try {
			return K(e)(t, n, r, i, a);
		} catch (e) {
			if (A(o), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function hr(e) {
		var t = j();
		try {
			K(e)();
		} catch (e) {
			if (A(t), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function gr(e, t, n, r, i, a, o, s, c, l, u) {
		var d = j();
		try {
			K(e)(t, n, r, i, a, o, s, c, l, u);
		} catch (e) {
			if (A(d), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function _r(e, t, n, r, i, a, o) {
		var s = j();
		try {
			K(e)(t, n, r, i, a, o);
		} catch (e) {
			if (A(s), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function vr(e, t, n, r, i, a, o, s, c, l, u, d, f, p, m, h, g) {
		var _ = j();
		try {
			K(e)(t, n, r, i, a, o, s, c, l, u, d, f, p, m, h, g);
		} catch (e) {
			if (A(_), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function yr(e, t, n, r, i, a, o, s, c) {
		var l = j();
		try {
			K(e)(t, n, r, i, a, o, s, c);
		} catch (e) {
			if (A(l), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function br(e, t, n, r, i) {
		var a = j();
		try {
			return K(e)(t, n, r, i);
		} catch (e) {
			if (A(a), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function xr(e, t, n, r, i, a, o, s) {
		var c = j();
		try {
			K(e)(t, n, r, i, a, o, s);
		} catch (e) {
			if (A(c), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Sr(e, t, n, r, i, a) {
		var o = j();
		try {
			return K(e)(t, n, r, i, a);
		} catch (e) {
			if (A(o), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Cr(e, t, n, r, i, a) {
		var o = j();
		try {
			K(e)(t, n, r, i, a);
		} catch (e) {
			if (A(o), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function wr(e, t, n, r, i, a, o) {
		var s = j();
		try {
			return K(e)(t, n, r, i, a, o);
		} catch (e) {
			if (A(s), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Tr(e, t, n, r, i, a, o, s) {
		var c = j();
		try {
			return K(e)(t, n, r, i, a, o, s);
		} catch (e) {
			if (A(c), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Er(e, t, n, r, i, a, o, s, c) {
		var l = j();
		try {
			return K(e)(t, n, r, i, a, o, s, c);
		} catch (e) {
			if (A(l), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Dr(e) {
		var t = j();
		try {
			return K(e)();
		} catch (e) {
			if (A(t), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Or(e, t, n) {
		var r = j();
		try {
			return K(e)(t, n);
		} catch (e) {
			if (A(r), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function kr(e, t, n, r, i, a, o) {
		var s = j();
		try {
			return K(e)(t, n, r, i, a, o);
		} catch (e) {
			if (A(s), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Ar(e, t, n, r) {
		var i = j();
		try {
			return K(e)(t, n, r);
		} catch (e) {
			if (A(i), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function jr(e, t, n, r) {
		var i = j();
		try {
			return K(e)(t, n, r);
		} catch (e) {
			if (A(i), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Mr(e, t, n, r, i, a, o, s, c) {
		var l = j();
		try {
			K(e)(t, n, r, i, a, o, s, c);
		} catch (e) {
			if (A(l), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Nr(e, t, n, r, i, a, o, s) {
		var c = j();
		try {
			K(e)(t, n, r, i, a, o, s);
		} catch (e) {
			if (A(c), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Pr(e, t, n, r, i, a, o, s, c, l) {
		var u = j();
		try {
			K(e)(t, n, r, i, a, o, s, c, l);
		} catch (e) {
			if (A(u), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Fr(e, t, n) {
		var r = j();
		try {
			return K(e)(t, n);
		} catch (e) {
			if (A(r), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Ir(e, t, n, r, i) {
		var a = j();
		try {
			return K(e)(t, n, r, i);
		} catch (e) {
			if (A(a), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Lr(e, t, n, r, i, a) {
		var o = j();
		try {
			return K(e)(t, n, r, i, a);
		} catch (e) {
			if (A(o), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Rr(e, t, n, r, i, a, o) {
		var s = j();
		try {
			K(e)(t, n, r, i, a, o);
		} catch (e) {
			if (A(s), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function zr(e, t, n, r) {
		var i = j();
		try {
			return K(e)(t, n, r);
		} catch (e) {
			if (A(i), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Br(e, t, n, r, i, a, o, s, c, l, u, d) {
		var f = j();
		try {
			return K(e)(t, n, r, i, a, o, s, c, l, u, d);
		} catch (e) {
			if (A(f), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Vr(e, t, n, r, i, a, o, s, c, l, u, d, f, p, m, h) {
		var g = j();
		try {
			K(e)(t, n, r, i, a, o, s, c, l, u, d, f, p, m, h);
		} catch (e) {
			if (A(g), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Hr(e, t, n, r, i, a, o, s, c, l, u, d, f, p, m, h, g, _, ee, v) {
		var y = j();
		try {
			er(e, t, n, r, i, a, o, s, c, l, u, d, f, p, m, h, g, _, ee, v);
		} catch (e) {
			if (A(y), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Ur(e, t, n, r, i, a, o) {
		var s = j();
		try {
			return tr(e, t, n, r, i, a, o);
		} catch (e) {
			if (A(s), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Wr(e, t, n, r, i) {
		var a = j();
		try {
			return nr(e, t, n, r, i);
		} catch (e) {
			if (A(a), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Gr() {
		ne();
		function e() {
			var e, t;
			i.calledRun = !0, !_ && (re(), (e = ee) == null || e(i), (t = i.onRuntimeInitialized) == null || t.call(i), ie());
		}
		i.setStatus ? (i.setStatus("Running..."), setTimeout(() => {
			setTimeout(() => i.setStatus(""), 1), e();
		}, 1)) : e();
	}
	var Kr = await fe();
	return Gr(), t = y ? i : new Promise((e, t) => {
		ee = e, v = t;
	}), t;
}
//#endregion
//#region src/full/index.ts
function T(e) {
	return m(w, e);
}
function E() {
	return oe(w);
}
function D(e) {
	return T({
		overrides: e,
		equalityFn: Object.is,
		fireImmediately: !0
	});
}
function O(e) {
	T({
		overrides: e,
		equalityFn: Object.is,
		fireImmediately: !1
	});
}
async function k(e, t) {
	return le(w, e, t);
}
async function pe(e, t) {
	return k(e, t);
}
async function me(e, t) {
	return k(e, t);
}
async function he(e, t) {
	return g(w, e, t);
}
var ge = "0f5ab81e45050ba10eed1b06689822a5d2042c7f780379dc0512b0f7366de275";
//#endregion
export { C as BARCODE_FORMATS, h as BARCODE_HRI_LABELS, S as BARCODE_META_FORMATS, d as BARCODE_SYMBOLOGIES, p as BINARIZERS, y as CHARACTER_SETS, se as CONTENT_TYPES, t as CREATABLE_BARCODE_FORMATS, _ as EAN_ADD_ON_SYMBOLS, x as GS1_BARCODE_FORMATS, f as INDUSTRIAL_BARCODE_FORMATS, r as LINEAR_BARCODE_FORMATS, n as MATRIX_BARCODE_FORMATS, l as READABLE_BARCODE_FORMATS, re as RETAIL_BARCODE_FORMATS, ie as TEXT_MODES, ce as ZXING_CPP_COMMIT, ge as ZXING_WASM_SHA256, te as ZXING_WASM_VERSION, e as barcodeFormats, fe as binarizers, v as characterSets, ae as contentTypes, b as defaultReaderOptions, ue as defaultWriterOptions, ee as eanAddOnSymbols, ne as encodeFormat, s as encodeFormats, c as formatToLabel, u as formatToSymbology, D as getZXingModule, i as linearBarcodeFormats, a as matrixBarcodeFormats, T as prepareZXingModule, E as purgeZXingModule, k as readBarcodes, me as readBarcodesFromImageData, pe as readBarcodesFromImageFile, O as setZXingModuleOverrides, o as symbologyToFormats, de as textModes, he as writeBarcode };
