// zxing-wasm/dist/es/reader/index.js
import { A as e, C as t, D as n, E as r, F as i, I as a, L as o, M as s, N as c, O as l, P as u, S as d, T as f, _ as p, a as m, b as h, d as g, f as _, g as v, h as ee, i as te, j as ne, k as re, l as ie, m as ae, n as oe, o as y, p as se, r as ce, s as le, u as ue, v as de, w as fe, x as pe, y as b } from "../share.js";
//#region src/reader/zxing_reader.js
async function x(e = {}) {
	var t, n, r, i = e, a = !!globalThis.window, o = typeof Bun < "u", s = !!globalThis.WorkerGlobalScope;
	!((n = globalThis.process) == null || (n = n.versions) == null) && n.node && ((r = globalThis.process) == null || r.type);
	var c = "./this.program", l, u = "";
	function d(e) {
		return i.locateFile ? i.locateFile(e, u) : u + e;
	}
	var f, p;
	if (a || s || o) {
		try {
			u = new URL(".", l).href;
		} catch {}
		s && (p = (e) => {
			var t = new XMLHttpRequest();
			return t.open("GET", e, !1), t.responseType = "arraybuffer", t.send(null), new Uint8Array(t.response);
		}), f = async (e) => {
			var t = await fetch(e, { credentials: "same-origin" });
			if (t.ok) return t.arrayBuffer();
			throw Error(t.status + " : " + t.url);
		};
	}
	var m = console.log.bind(console), h = console.error.bind(console), g, _ = !1, v, ee, te = !1;
	function ne() {
		var e = Xn.buffer;
		S = new Int8Array(e), b = new Int16Array(e), i.HEAPU8 = T = new Uint8Array(e), C = new Uint16Array(e), x = new Int32Array(e), w = new Uint32Array(e), me = new Float32Array(e), he = new Float64Array(e);
	}
	function re() {
		if (i.preRun) for (typeof i.preRun == "function" && (i.preRun = [i.preRun]); i.preRun.length;) be(i.preRun.shift());
		ge(ye);
	}
	function ie() {
		te = !0, zr.Ba();
	}
	function ae() {
		if (i.postRun) for (typeof i.postRun == "function" && (i.postRun = [i.postRun]); i.postRun.length;) ve(i.postRun.shift());
		ge(_e);
	}
	function oe(e) {
		var t, n;
		(t = i.onAbort) == null || t.call(i, e), e = "Aborted(" + e + ")", h(e), _ = !0, e += ". Build with -sASSERTIONS for more info.";
		var r = new WebAssembly.RuntimeError(e);
		throw (n = ee) == null || n(r), r;
	}
	var y;
	function se() {
		return d("zxing_reader.wasm");
	}
	function ce(e) {
		if (e == y && g) return new Uint8Array(g);
		if (p) return p(e);
		throw "both async and sync fetching of the wasm failed";
	}
	async function le(e) {
		if (!g) try {
			var t = await f(e);
			return new Uint8Array(t);
		} catch {}
		return ce(e);
	}
	async function ue(e, t) {
		try {
			var n = await le(e);
			return await WebAssembly.instantiate(n, t);
		} catch (e) {
			h(`failed to asynchronously prepare wasm: ${e}`), oe(e);
		}
	}
	async function de(e, t, n) {
		if (!e && WebAssembly.instantiateStreaming) try {
			var r = fetch(t, { credentials: "same-origin" });
			return await WebAssembly.instantiateStreaming(r, n);
		} catch (e) {
			h(`wasm streaming compile failed: ${e}`), h("falling back to ArrayBuffer instantiation");
		}
		return ue(t, n);
	}
	function fe() {
		return { a: $n };
	}
	async function pe() {
		function e(e, t) {
			return zr = e.exports, Qn(zr), ne(), zr;
		}
		function t(t) {
			return e(t.instance);
		}
		var n = fe();
		return i.instantiateWasm ? new Promise((t, r) => {
			i.instantiateWasm(n, (n, r) => {
				t(e(n, r));
			});
		}) : (y != null || (y = se()), t(await de(g, y, n)));
	}
	var b, x, S, me, he, C, w, T, ge = (e) => {
		for (; e.length > 0;) e.shift()(i);
	}, _e = [], ve = (e) => _e.push(e), ye = [], be = (e) => ye.push(e), E = (e) => Hn(e), D = () => Un(), xe = [], O = 0, Se = (e) => {
		var t = new we(e);
		return t.get_caught() || (t.set_caught(!0), O--), t.set_rethrown(!1), xe.push(t), Bn(e);
	}, k = 0, Ce = () => {
		$(0, 0);
		var e = xe.pop();
		Wn(e.excPtr), k = 0;
	};
	class we {
		constructor(e) {
			this.excPtr = e, this.ptr = e - 24;
		}
		set_type(e) {
			w[this.ptr + 4 >> 2] = e;
		}
		get_type() {
			return w[this.ptr + 4 >> 2];
		}
		set_destructor(e) {
			w[this.ptr + 8 >> 2] = e;
		}
		get_destructor() {
			return w[this.ptr + 8 >> 2];
		}
		set_caught(e) {
			e = +!!e, S[this.ptr + 12] = e;
		}
		get_caught() {
			return S[this.ptr + 12] != 0;
		}
		set_rethrown(e) {
			e = +!!e, S[this.ptr + 13] = e;
		}
		get_rethrown() {
			return S[this.ptr + 13] != 0;
		}
		init(e, t) {
			this.set_adjusted_ptr(0), this.set_type(e), this.set_destructor(t);
		}
		set_adjusted_ptr(e) {
			w[this.ptr + 16 >> 2] = e;
		}
		get_adjusted_ptr() {
			return w[this.ptr + 16 >> 2];
		}
	}
	var A = (e) => Vn(e), Te = (e) => {
		var t = k;
		if (!t) return A(0), 0;
		var n = new we(t);
		n.set_adjusted_ptr(t);
		var r = n.get_type();
		if (!r) return A(0), t;
		for (var i of e) {
			if (i === 0 || i === r) break;
			var a = n.ptr + 16;
			if (Kn(i, r, a)) return A(i), t;
		}
		return A(r), t;
	}, Ee = () => Te([]), De = (e) => Te([e]), Oe = (e, t) => Te([e, t]), ke = () => {
		var e = xe.pop();
		e || oe("no exception to throw");
		var t = e.excPtr;
		throw e.get_rethrown() || (xe.push(e), e.set_rethrown(!0), e.set_caught(!1), O++), Gn(t), k = t, k;
	}, Ae = (e, t, n) => {
		throw new we(e).init(t, n), Gn(e), k = e, O++, k;
	}, je = () => O, Me = (e) => {
		throw k || (k = e), k;
	}, Ne = () => oe(""), j = {}, Pe = (e) => {
		for (; e.length;) {
			var t = e.pop();
			e.pop()(t);
		}
	};
	function M(e) {
		return this.fromWireType(w[e >> 2]);
	}
	var N = {}, P = {}, F = {}, Fe = class extends Error {
		constructor(e) {
			super(e), this.name = "InternalError";
		}
	}, I = (e) => {
		throw new Fe(e);
	}, L = (e, t, n) => {
		e.forEach((e) => F[e] = t);
		function r(t) {
			var r = n(t);
			r.length !== e.length && I("Mismatched type converter count");
			for (var i = 0; i < e.length; ++i) V(e[i], r[i]);
		}
		var i = Array(t.length), a = [], o = 0;
		{
			let e = t;
			for (let t = 0; t < e.length; ++t) {
				let n = e[t];
				P.hasOwnProperty(n) ? i[t] = P[n] : (a.push(n), N.hasOwnProperty(n) || (N[n] = []), N[n].push(() => {
					i[t] = P[n], ++o, o === a.length && r(i);
				}));
			}
		}
		a.length === 0 && r(i);
	}, Ie = (e) => {
		var t = j[e];
		delete j[e];
		var n = t.rawConstructor, r = t.rawDestructor, i = t.fields, a = i.map((e) => e.getterReturnType).concat(i.map((e) => e.setterArgumentType));
		L([e], a, (e) => {
			var a = {};
			{
				let t = i;
				for (let n = 0; n < t.length; ++n) {
					let r = t[n], o = e[n], s = r.getter, c = r.getterContext, l = e[n + i.length], u = r.setter, d = r.setterContext;
					a[r.fieldName] = {
						read: (e) => o.fromWireType(s(c, e)),
						write: (e, t) => {
							var n = [];
							u(d, e, l.toWireType(n, t)), Pe(n);
						},
						optional: o.optional
					};
				}
			}
			return [{
				name: t.name,
				fromWireType: (e) => {
					var t = {};
					for (var n in a) t[n] = a[n].read(e);
					return r(e), t;
				},
				toWireType: (e, t) => {
					for (var i in a) if (!(i in t) && !a[i].optional) throw TypeError(`Missing field: "${i}"`);
					var o = n();
					for (i in a) a[i].write(o, t[i]);
					return e !== null && e.push(r, o), o;
				},
				readValueFromPointer: M,
				destructorFunction: r
			}];
		});
	}, Le = (e, t, n, r, i) => {}, R = (e) => {
		for (var t = "";;) {
			var n = T[e++];
			if (!n) return t;
			t += String.fromCharCode(n);
		}
	}, z = class extends Error {
		constructor(e) {
			super(e), this.name = "BindingError";
		}
	}, B = (e) => {
		throw new z(e);
	};
	function Re(e, t) {
		let n = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {};
		var r = t.name;
		if (e || B(`type "${r}" must have a positive integer typeid pointer`), P.hasOwnProperty(e)) {
			if (n.ignoreDuplicateRegistrations) return;
			B(`Cannot register type '${r}' twice`);
		}
		if (P[e] = t, delete F[e], N.hasOwnProperty(e)) {
			var i = N[e];
			delete N[e], i.forEach((e) => e());
		}
	}
	function V(e, t) {
		return Re(e, t, arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {});
	}
	var ze = (e, t, n, r) => {
		t = R(t), V(e, {
			name: t,
			fromWireType: function(e) {
				return !!e;
			},
			toWireType: function(e, t) {
				return t ? n : r;
			},
			readValueFromPointer: function(e) {
				return this.fromWireType(T[e]);
			},
			destructorFunction: null
		});
	}, Be = (e) => ({
		count: e.count,
		deleteScheduled: e.deleteScheduled,
		preservePointerOnDelete: e.preservePointerOnDelete,
		ptr: e.ptr,
		ptrType: e.ptrType,
		smartPtr: e.smartPtr,
		smartPtrType: e.smartPtrType
	}), Ve = (e) => {
		function t(e) {
			return e.$$.ptrType.registeredClass.name;
		}
		B(t(e) + " instance already deleted");
	}, He = !1, Ue = (e) => {}, We = (e) => {
		e.smartPtr ? e.smartPtrType.rawDestructor(e.smartPtr) : e.ptrType.registeredClass.rawDestructor(e.ptr);
	}, Ge = (e) => {
		--e.count.value, e.count.value === 0 && We(e);
	}, H = (e) => globalThis.FinalizationRegistry ? (He = new FinalizationRegistry((e) => {
		Ge(e.$$);
	}), H = (e) => {
		var t = e.$$;
		if (t.smartPtr) {
			var n = { $$: t };
			He.register(e, n, e);
		}
		return e;
	}, Ue = (e) => He.unregister(e), H(e)) : (H = (e) => e, e), U = [], Ke = () => {
		for (; U.length;) {
			var e = U.pop();
			e.$$.deleteScheduled = !1, e.delete();
		}
	}, qe, Je = () => {
		let e = W.prototype;
		Object.assign(e, {
			isAliasOf(e) {
				if (!(this instanceof W) || !(e instanceof W)) return !1;
				var t = this.$$.ptrType.registeredClass, n = this.$$.ptr;
				e.$$ = e.$$;
				for (var r = e.$$.ptrType.registeredClass, i = e.$$.ptr; t.baseClass;) n = t.upcast(n), t = t.baseClass;
				for (; r.baseClass;) i = r.upcast(i), r = r.baseClass;
				return t === r && n === i;
			},
			clone() {
				if (this.$$.ptr || Ve(this), this.$$.preservePointerOnDelete) return this.$$.count.value += 1, this;
				var e = H(Object.create(Object.getPrototypeOf(this), { $$: { value: Be(this.$$) } }));
				return e.$$.count.value += 1, e.$$.deleteScheduled = !1, e;
			},
			delete() {
				this.$$.ptr || Ve(this), this.$$.deleteScheduled && !this.$$.preservePointerOnDelete && B("Object already scheduled for deletion"), Ue(this), Ge(this.$$), this.$$.preservePointerOnDelete || (this.$$.smartPtr = void 0, this.$$.ptr = void 0);
			},
			isDeleted() {
				return !this.$$.ptr;
			},
			deleteLater() {
				return this.$$.ptr || Ve(this), this.$$.deleteScheduled && !this.$$.preservePointerOnDelete && B("Object already scheduled for deletion"), U.push(this), U.length === 1 && qe && qe(Ke), this.$$.deleteScheduled = !0, this;
			}
		});
		let t = Symbol.dispose;
		t && (e[t] = e.delete);
	};
	function W() {}
	var Ye = (e, t) => Object.defineProperty(t, "name", { value: e }), Xe = {}, Ze = (e, t, n) => {
		if (e[t].overloadTable === void 0) {
			var r = e[t];
			e[t] = function() {
				var r = [...arguments];
				return e[t].overloadTable.hasOwnProperty(r.length) || B(`Function '${n}' called with an invalid number of arguments (${r.length}) - expects one of (${e[t].overloadTable})!`), e[t].overloadTable[r.length].apply(this, r);
			}, e[t].overloadTable = [], e[t].overloadTable[r.argCount] = r;
		}
	}, Qe = (e, t, n) => {
		i.hasOwnProperty(e) ? ((n === void 0 || i[e].overloadTable !== void 0 && i[e].overloadTable[n] !== void 0) && B(`Cannot register public name '${e}' twice`), Ze(i, e, e), i[e].overloadTable.hasOwnProperty(n) && B(`Cannot register multiple overloads of a function with the same number of arguments (${n})!`), i[e].overloadTable[n] = t) : (i[e] = t, i[e].argCount = n);
	}, $e = 48, et = 57, tt = (e) => {
		e = e.replace(/[^a-zA-Z0-9_]/g, "$");
		var t = e.charCodeAt(0);
		return t >= $e && t <= et ? `_${e}` : e;
	};
	function nt(e, t, n, r, i, a, o, s) {
		this.name = e, this.constructor = t, this.instancePrototype = n, this.rawDestructor = r, this.baseClass = i, this.getActualType = a, this.upcast = o, this.downcast = s, this.pureVirtualFunctions = [];
	}
	var rt = (e, t, n) => {
		for (; t !== n;) t.upcast || B(`Expected null or instance of ${n.name}, got an instance of ${t.name}`), e = t.upcast(e), t = t.baseClass;
		return e;
	}, it = (e) => {
		if (e === null) return "null";
		var t = typeof e;
		return t === "object" || t === "array" || t === "function" ? e.toString() : "" + e;
	};
	function at(e, t) {
		if (t === null) return this.isReference && B(`null is not a valid ${this.name}`), 0;
		t.$$ || B(`Cannot pass "${it(t)}" as a ${this.name}`), t.$$.ptr || B(`Cannot pass deleted object as a pointer of type ${this.name}`);
		var n = t.$$.ptrType.registeredClass;
		return rt(t.$$.ptr, n, this.registeredClass);
	}
	function ot(e, t) {
		var n;
		if (t === null) return this.isReference && B(`null is not a valid ${this.name}`), this.isSmartPointer ? (n = this.rawConstructor(), e !== null && e.push(this.rawDestructor, n), n) : 0;
		(!t || !t.$$) && B(`Cannot pass "${it(t)}" as a ${this.name}`), t.$$.ptr || B(`Cannot pass deleted object as a pointer of type ${this.name}`), !this.isConst && t.$$.ptrType.isConst && B(`Cannot convert argument of type ${t.$$.smartPtrType ? t.$$.smartPtrType.name : t.$$.ptrType.name} to parameter type ${this.name}`);
		var r = t.$$.ptrType.registeredClass;
		if (n = rt(t.$$.ptr, r, this.registeredClass), this.isSmartPointer) switch (t.$$.smartPtr === void 0 && B("Passing raw pointer to smart pointer is illegal"), this.sharingPolicy) {
			case 0:
				t.$$.smartPtrType === this ? n = t.$$.smartPtr : B(`Cannot convert argument of type ${t.$$.smartPtrType ? t.$$.smartPtrType.name : t.$$.ptrType.name} to parameter type ${this.name}`);
				break;
			case 1:
				n = t.$$.smartPtr;
				break;
			case 2:
				if (t.$$.smartPtrType === this) n = t.$$.smartPtr;
				else {
					var i = t.clone();
					n = this.rawShare(n, Y.toHandle(() => i.delete())), e !== null && e.push(this.rawDestructor, n);
				}
				break;
			default: B("Unsupported sharing policy");
		}
		return n;
	}
	function st(e, t) {
		if (t === null) return this.isReference && B(`null is not a valid ${this.name}`), 0;
		t.$$ || B(`Cannot pass "${it(t)}" as a ${this.name}`), t.$$.ptr || B(`Cannot pass deleted object as a pointer of type ${this.name}`), t.$$.ptrType.isConst && B(`Cannot convert argument of type ${t.$$.ptrType.name} to parameter type ${this.name}`);
		var n = t.$$.ptrType.registeredClass;
		return rt(t.$$.ptr, n, this.registeredClass);
	}
	var ct = (e, t, n) => {
		if (t === n) return e;
		if (n.baseClass === void 0) return null;
		var r = ct(e, t, n.baseClass);
		return r === null ? null : n.downcast(r);
	}, lt = {}, ut = (e, t) => {
		for (t === void 0 && B("ptr should not be undefined"); e.baseClass;) t = e.upcast(t), e = e.baseClass;
		return t;
	}, dt = (e, t) => (t = ut(e, t), lt[t]), ft = (e, t) => ((!t.ptrType || !t.ptr) && I("makeClassHandle requires ptr and ptrType"), !!t.smartPtrType != !!t.smartPtr && I("Both smartPtrType and smartPtr must be specified"), t.count = { value: 1 }, H(Object.create(e, { $$: {
		value: t,
		writable: !0
	} })));
	function pt(e) {
		var t = this.getPointee(e);
		if (!t) return this.destructor(e), null;
		var n = dt(this.registeredClass, t);
		if (n !== void 0) {
			if (n.$$.count.value === 0) return n.$$.ptr = t, n.$$.smartPtr = e, n.clone();
			var r = n.clone();
			return this.destructor(e), r;
		}
		function i() {
			return this.isSmartPointer ? ft(this.registeredClass.instancePrototype, {
				ptrType: this.pointeeType,
				ptr: t,
				smartPtrType: this,
				smartPtr: e
			}) : ft(this.registeredClass.instancePrototype, {
				ptrType: this,
				ptr: e
			});
		}
		var a = Xe[this.registeredClass.getActualType(t)];
		if (!a) return i.call(this);
		var o = this.isConst ? a.constPointerType : a.pointerType, s = ct(t, this.registeredClass, o.registeredClass);
		return s === null ? i.call(this) : this.isSmartPointer ? ft(o.registeredClass.instancePrototype, {
			ptrType: o,
			ptr: s,
			smartPtrType: this,
			smartPtr: e
		}) : ft(o.registeredClass.instancePrototype, {
			ptrType: o,
			ptr: s
		});
	}
	var mt = () => {
		Object.assign(ht.prototype, {
			getPointee(e) {
				return this.rawGetPointee && (e = this.rawGetPointee(e)), e;
			},
			destructor(e) {
				var t;
				(t = this.rawDestructor) == null || t.call(this, e);
			},
			readValueFromPointer: M,
			fromWireType: pt
		});
	};
	function ht(e, t, n, r, i, a, o, s, c, l, u) {
		this.name = e, this.registeredClass = t, this.isReference = n, this.isConst = r, this.isSmartPointer = i, this.pointeeType = a, this.sharingPolicy = o, this.rawGetPointee = s, this.rawConstructor = c, this.rawShare = l, this.rawDestructor = u, !i && t.baseClass === void 0 ? r ? (this.toWireType = at, this.destructorFunction = null) : (this.toWireType = st, this.destructorFunction = null) : this.toWireType = ot;
	}
	var gt = (e, t, n) => {
		i.hasOwnProperty(e) || I("Replacing nonexistent public symbol"), i[e].overloadTable !== void 0 && n !== void 0 ? i[e].overloadTable[n] = t : (i[e] = t, i[e].argCount = n);
	}, G = {}, _t = (e, t, n) => {
		e = e.replace(/p/g, "i");
		var r = G[e];
		return r(t, ...n);
	}, vt = [], K = (e) => {
		var t = vt[e];
		return t || (vt[e] = t = Zn.get(e)), t;
	}, yt = function(e, t) {
		let n = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : [];
		if (arguments.length > 3 && arguments[3] !== void 0 && arguments[3], e.includes("j")) return _t(e, t, n);
		var r = K(t)(...n);
		function i(e) {
			return e;
		}
		return i(r);
	}, bt = function(e, t) {
		let n = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : !1;
		return function() {
			return yt(e, t, [...arguments], n);
		};
	}, q = function(e, t) {
		arguments.length > 2 && arguments[2] !== void 0 && arguments[2], e = R(e);
		function n() {
			return e.includes("j") ? bt(e, t) : K(t);
		}
		var r = n();
		return typeof r != "function" && B(`unknown function pointer with signature ${e}: ${t}`), r;
	};
	class xt extends Error {}
	var St = (e) => {
		var t = Rn(e), n = R(t);
		return Q(t), n;
	}, Ct = (e, t) => {
		var n = [], r = {};
		function i(e) {
			if (!r[e] && !P[e]) {
				if (F[e]) {
					F[e].forEach(i);
					return;
				}
				n.push(e), r[e] = !0;
			}
		}
		throw t.forEach(i), new xt(`${e}: ` + n.map(St).join([", "]));
	}, wt = (e, t, n, r, i, a, o, s, c, l, u, d, f) => {
		u = R(u), a = q(i, a), s && (s = q(o, s)), l && (l = q(c, l)), f = q(d, f);
		var p = tt(u);
		Qe(p, function() {
			Ct(`Cannot construct ${u} due to unbound types`, [r]);
		}), L([
			e,
			t,
			n
		], r ? [r] : [], (t) => {
			t = t[0];
			var n, i;
			r ? (n = t.registeredClass, i = n.instancePrototype) : i = W.prototype;
			var o = Ye(u, function() {
				if (Object.getPrototypeOf(this) !== c) throw new z(`Use 'new' to construct ${u}`);
				if (d.constructor_body === void 0) throw new z(`${u} has no accessible constructor`);
				var e = [...arguments], t = d.constructor_body[e.length];
				if (t === void 0) throw new z(`Tried to invoke ctor of ${u} with invalid number of parameters (${e.length}) - expected (${Object.keys(d.constructor_body).toString()}) parameters instead!`);
				return t.apply(this, e);
			}), c = Object.create(i, { constructor: { value: o } });
			o.prototype = c;
			var d = new nt(u, o, c, f, n, a, s, l);
			if (d.baseClass) {
				var m;
				(m = d.baseClass).__derivedClasses != null || (m.__derivedClasses = []), d.baseClass.__derivedClasses.push(d);
			}
			var h = new ht(u, d, !0, !1, !1), g = new ht(u + "*", d, !1, !1, !1), _ = new ht(u + " const*", d, !1, !0, !1);
			return Xe[e] = {
				pointerType: g,
				constPointerType: _
			}, gt(p, o), [
				h,
				g,
				_
			];
		});
	}, Tt = (e, t) => {
		for (var n = [], r = 0; r < e; r++) n.push(w[t + r * 4 >> 2]);
		return n;
	};
	function Et(e) {
		for (var t = 1; t < e.length; ++t) if (e[t] !== null && e[t].destructorFunction === void 0) return !0;
		return !1;
	}
	function Dt(e, t, n, r, i, a) {
		var o = t.length;
		o < 2 && B("argTypes array size mismatch! Must at least get return value and 'this' types!");
		var s = t[1] !== null && n !== null, c = Et(t), l = !t[0].isVoid, u = o - 2, d = Array(u), f = [], p = [];
		return Ye(e, function() {
			p.length = 0;
			var e;
			f.length = s ? 2 : 1, f[0] = i, s && (e = t[1].toWireType(p, this), f[1] = e);
			for (var n = 0; n < u; ++n) d[n] = t[n + 2].toWireType(p, n < 0 || arguments.length <= n ? void 0 : arguments[n]), f.push(d[n]);
			var a = r(...f);
			function o(n) {
				if (c) Pe(p);
				else for (var r = s ? 1 : 2; r < t.length; r++) {
					var i = r === 1 ? e : d[r - 2];
					t[r].destructorFunction !== null && t[r].destructorFunction(i);
				}
				if (l) return t[0].fromWireType(n);
			}
			return o(a);
		});
	}
	var Ot = (e, t, n, r, i, a) => {
		var o = Tt(t, n);
		i = q(r, i), L([], [e], (e) => {
			e = e[0];
			var n = `constructor ${e.name}`;
			if (e.registeredClass.constructor_body === void 0 && (e.registeredClass.constructor_body = []), e.registeredClass.constructor_body[t - 1] !== void 0) throw new z(`Cannot register multiple constructors with identical number of parameters (${t - 1}) for class '${e.name}'! Overload resolution is currently only performed using the parameter count, not actual type info!`);
			return e.registeredClass.constructor_body[t - 1] = () => {
				Ct(`Cannot construct ${e.name} due to unbound types`, o);
			}, L([], o, (r) => (r.splice(1, 0, null), e.registeredClass.constructor_body[t - 1] = Dt(n, r, null, i, a), [])), [];
		});
	}, kt = (e) => {
		e = e.trim();
		let t = e.indexOf("(");
		return t === -1 ? e : e.slice(0, t);
	}, At = (e, t, n, r, i, a, o, s, c, l) => {
		var u = Tt(n, r);
		t = R(t), t = kt(t), a = q(i, a, c), L([], [e], (e) => {
			e = e[0];
			var r = `${e.name}.${t}`;
			t.startsWith("@@") && (t = Symbol[t.substring(2)]), s && e.registeredClass.pureVirtualFunctions.push(t);
			function i() {
				Ct(`Cannot call ${r} due to unbound types`, u);
			}
			var l = e.registeredClass.instancePrototype, d = l[t];
			return d === void 0 || d.overloadTable === void 0 && d.className !== e.name && d.argCount === n - 2 ? (i.argCount = n - 2, i.className = e.name, l[t] = i) : (Ze(l, t, r), l[t].overloadTable[n - 2] = i), L([], u, (i) => {
				var s = Dt(r, i, e, a, o, c);
				return l[t].overloadTable === void 0 ? (s.argCount = n - 2, l[t] = s) : l[t].overloadTable[n - 2] = s, [];
			}), [];
		});
	}, jt = [], J = [
		0,
		1,
		,
		1,
		null,
		1,
		!0,
		1,
		!1,
		1
	], Mt = (e) => {
		e > 9 && --J[e + 1] === 0 && (J[e] = void 0, jt.push(e));
	}, Y = {
		toValue: (e) => (e || B(`Cannot use deleted val. handle = ${e}`), J[e]),
		toHandle: (e) => {
			switch (e) {
				case void 0: return 2;
				case null: return 4;
				case !0: return 6;
				case !1: return 8;
				default: {
					let t = jt.pop() || J.length;
					return J[t] = e, J[t + 1] = 1, t;
				}
			}
		}
	}, Nt = {
		name: "emscripten::val",
		fromWireType: (e) => {
			var t = Y.toValue(e);
			return Mt(e), t;
		},
		toWireType: (e, t) => Y.toHandle(t),
		readValueFromPointer: M,
		destructorFunction: null
	}, Pt = (e) => V(e, Nt), Ft = (e, t) => {
		switch (t) {
			case 4: return function(e) {
				return this.fromWireType(me[e >> 2]);
			};
			case 8: return function(e) {
				return this.fromWireType(he[e >> 3]);
			};
			default: throw TypeError(`invalid float width (${t}): ${e}`);
		}
	}, It = (e, t, n) => {
		t = R(t), V(e, {
			name: t,
			fromWireType: (e) => e,
			toWireType: (e, t) => t,
			readValueFromPointer: Ft(t, n),
			destructorFunction: null
		});
	}, Lt = (e, t, n, r, i, a, o, s) => {
		var c = Tt(t, n);
		e = R(e), e = kt(e), i = q(r, i, o), Qe(e, function() {
			Ct(`Cannot call ${e} due to unbound types`, c);
		}, t - 1), L([], c, (n) => {
			var r = [n[0], null].concat(n.slice(1));
			return gt(e, Dt(e, r, null, i, a, o), t - 1), [];
		});
	}, Rt = (e, t, n) => {
		switch (t) {
			case 1: return n ? (e) => S[e] : (e) => T[e];
			case 2: return n ? (e) => b[e >> 1] : (e) => C[e >> 1];
			case 4: return n ? (e) => x[e >> 2] : (e) => w[e >> 2];
			default: throw TypeError(`invalid integer width (${t}): ${e}`);
		}
	}, zt = (e, t, n, r, i) => {
		t = R(t);
		let a = r === 0, o = (e) => e;
		if (a) {
			var s = 32 - 8 * n;
			o = (e) => e << s >>> s, i = o(i);
		}
		V(e, {
			name: t,
			fromWireType: o,
			toWireType: (e, t) => t,
			readValueFromPointer: Rt(t, n, r !== 0),
			destructorFunction: null
		});
	}, Bt = (e, t, n) => {
		let r = (e, t) => {
			let n = 0;
			return {
				next() {
					if (n >= e) return { done: !0 };
					let r = n;
					return n++, {
						value: t(r),
						done: !1
					};
				},
				[Symbol.iterator]() {
					return this;
				}
			};
		};
		e[Symbol.iterator] || (e[Symbol.iterator] = function() {
			return r(this[t](), (e) => this[n](e));
		});
	}, Vt = (e, t, n, r) => {
		n = R(n), r = R(r), L([], [e, t], (e) => {
			let t = e[0];
			return Bt(t.registeredClass.instancePrototype, n, r), [];
		});
	}, Ht = (e, t, n) => {
		var r = [
			Int8Array,
			Uint8Array,
			Int16Array,
			Uint16Array,
			Int32Array,
			Uint32Array,
			Float32Array,
			Float64Array
		][t];
		function i(e) {
			var t = w[e >> 2], n = w[e + 4 >> 2];
			return new r(S.buffer, n, t);
		}
		n = R(n), V(e, {
			name: n,
			fromWireType: i,
			readValueFromPointer: i
		}, { ignoreDuplicateRegistrations: !0 });
	}, Ut = Object.assign({ optional: !0 }, Nt), Wt = (e, t) => {
		V(e, Ut);
	}, Gt = (e, t, n, r) => {
		if (!(r > 0)) return 0;
		for (var i = n, a = n + r - 1, o = 0; o < e.length; ++o) {
			var s = e.codePointAt(o);
			if (s <= 127) {
				if (n >= a) break;
				t[n++] = s;
			} else if (s <= 2047) {
				if (n + 1 >= a) break;
				t[n++] = 192 | s >> 6, t[n++] = 128 | s & 63;
			} else if (s <= 65535) {
				if (n + 2 >= a) break;
				t[n++] = 224 | s >> 12, t[n++] = 128 | s >> 6 & 63, t[n++] = 128 | s & 63;
			} else {
				if (n + 3 >= a) break;
				t[n++] = 240 | s >> 18, t[n++] = 128 | s >> 12 & 63, t[n++] = 128 | s >> 6 & 63, t[n++] = 128 | s & 63, o++;
			}
		}
		return t[n] = 0, n - i;
	}, X = (e, t, n) => Gt(e, T, t, n), Kt = (e) => {
		for (var t = 0, n = 0; n < e.length; ++n) {
			var r = e.charCodeAt(n);
			r <= 127 ? t++ : r <= 2047 ? t += 2 : r >= 55296 && r <= 57343 ? (t += 4, ++n) : t += 3;
		}
		return t;
	}, qt = globalThis.TextDecoder && new TextDecoder(), Jt = (e, t, n, r) => {
		var i = t + n;
		if (r) return i;
		for (; e[t] && !(t >= i);) ++t;
		return t;
	}, Yt = function(e) {
		let t = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : 0, n = arguments.length > 2 ? arguments[2] : void 0, r = arguments.length > 3 ? arguments[3] : void 0;
		var i = Jt(e, t, n, r);
		if (i - t > 16 && e.buffer && qt) return qt.decode(e.subarray(t, i));
		for (var a = ""; t < i;) {
			var o = e[t++];
			if (!(o & 128)) {
				a += String.fromCharCode(o);
				continue;
			}
			var s = e[t++] & 63;
			if ((o & 224) == 192) {
				a += String.fromCharCode((o & 31) << 6 | s);
				continue;
			}
			var c = e[t++] & 63;
			if (o = (o & 240) == 224 ? (o & 15) << 12 | s << 6 | c : (o & 7) << 18 | s << 12 | c << 6 | e[t++] & 63, o < 65536) a += String.fromCharCode(o);
			else {
				var l = o - 65536;
				a += String.fromCharCode(55296 | l >> 10, 56320 | l & 1023);
			}
		}
		return a;
	}, Xt = (e, t, n) => e ? Yt(T, e, t, n) : "", Zt = (e, t) => {
		t = R(t);
		var n = !0;
		V(e, {
			name: t,
			fromWireType(e) {
				var t = w[e >> 2], r = e + 4, i;
				if (n) i = Xt(r, t, !0);
				else {
					i = "";
					for (var a = 0; a < t; ++a) i += String.fromCharCode(T[r + a]);
				}
				return Q(e), i;
			},
			toWireType(e, t) {
				t instanceof ArrayBuffer && (t = new Uint8Array(t));
				var r, i = typeof t == "string";
				i || ArrayBuffer.isView(t) && t.BYTES_PER_ELEMENT == 1 || B("Cannot pass non-string to std::string"), r = n && i ? Kt(t) : t.length;
				var a = zn(4 + r + 1), o = a + 4;
				if (w[a >> 2] = r, i) if (n) X(t, o, r + 1);
				else for (var s = 0; s < r; ++s) {
					var c = t.charCodeAt(s);
					c > 255 && (Q(a), B("String has UTF-16 code units that do not fit in 8 bits")), T[o + s] = c;
				}
				else T.set(t, o);
				return e !== null && e.push(Q, a), a;
			},
			readValueFromPointer: M,
			destructorFunction(e) {
				Q(e);
			}
		});
	}, Qt = globalThis.TextDecoder ? new TextDecoder("utf-16le") : void 0, $t = (e, t, n) => {
		var r = e >> 1, i = Jt(C, r, t / 2, n);
		if (i - r > 16 && Qt) return Qt.decode(C.subarray(r, i));
		for (var a = "", o = r; o < i; ++o) {
			var s = C[o];
			a += String.fromCharCode(s);
		}
		return a;
	}, en = (e, t, n) => {
		if (n != null || (n = 2147483647), n < 2) return 0;
		n -= 2;
		for (var r = t, i = n < e.length * 2 ? n / 2 : e.length, a = 0; a < i; ++a) {
			var o = e.charCodeAt(a);
			b[t >> 1] = o, t += 2;
		}
		return b[t >> 1] = 0, t - r;
	}, tn = (e) => e.length * 2, nn = (e, t, n) => {
		for (var r = "", i = e >> 2, a = 0; !(a >= t / 4); a++) {
			var o = w[i + a];
			if (!o && !n) break;
			r += String.fromCodePoint(o);
		}
		return r;
	}, rn = (e, t, n) => {
		if (n != null || (n = 2147483647), n < 4) return 0;
		for (var r = t, i = r + n - 4, a = 0; a < e.length; ++a) {
			var o = e.codePointAt(a);
			if (o > 65535 && a++, x[t >> 2] = o, t += 4, t + 4 > i) break;
		}
		return x[t >> 2] = 0, t - r;
	}, an = (e) => {
		for (var t = 0, n = 0; n < e.length; ++n) e.codePointAt(n) > 65535 && n++, t += 4;
		return t;
	}, on = (e, t, n) => {
		n = R(n);
		var r, i, a;
		t === 2 ? (r = $t, i = en, a = tn) : (r = nn, i = rn, a = an), V(e, {
			name: n,
			fromWireType: (e) => {
				var n = w[e >> 2], i = r(e + 4, n * t, !0);
				return Q(e), i;
			},
			toWireType: (e, r) => {
				typeof r != "string" && B(`Cannot pass non-string to C++ string type ${n}`);
				var o = a(r), s = zn(4 + o + t);
				return w[s >> 2] = o / t, i(r, s + 4, o + t), e !== null && e.push(Q, s), s;
			},
			readValueFromPointer: M,
			destructorFunction(e) {
				Q(e);
			}
		});
	}, sn = (e, t, n, r, i, a) => {
		j[e] = {
			name: R(t),
			rawConstructor: q(n, r),
			rawDestructor: q(i, a),
			fields: []
		};
	}, cn = (e, t, n, r, i, a, o, s, c, l) => {
		j[e].fields.push({
			fieldName: R(t),
			getterReturnType: n,
			getter: q(r, i),
			getterContext: a,
			setterArgumentType: o,
			setter: q(s, c),
			setterContext: l
		});
	}, ln = (e, t) => {
		t = R(t), V(e, {
			isVoid: !0,
			name: t,
			fromWireType: () => void 0,
			toWireType: (e, t) => void 0
		});
	}, un = [], dn = (e) => {
		var t = un.length;
		return un.push(e), t;
	}, fn = (e, t) => {
		var n = P[e];
		return n === void 0 && B(`${t} has unknown type ${St(e)}`), n;
	}, pn = (e, t) => {
		for (var n = Array(e), r = 0; r < e; ++r) n[r] = fn(w[t + r * 4 >> 2], `parameter ${r}`);
		return n;
	}, mn = (e, t, n) => {
		var r = [], i = e(r, n);
		return r.length && (w[t >> 2] = Y.toHandle(r)), i;
	}, hn = {}, gn = (e) => {
		var t = hn[e];
		return t === void 0 ? R(e) : t;
	}, _n = (e, t, n) => {
		var r = 8, [i, ...a] = pn(e, t), o = i.toWireType.bind(i), s = a.map((e) => e.readValueFromPointer.bind(e));
		e--;
		var c = Array(e);
		return dn(Ye(`methodCaller<(${a.map((e) => e.name)}) => ${i.name}>`, (t, i, a, l) => {
			for (var u = 0, d = 0; d < e; ++d) c[d] = s[d](l + u), u += r;
			var f;
			switch (n) {
				case 0:
					f = Y.toValue(t).apply(null, c);
					break;
				case 2:
					f = Reflect.construct(Y.toValue(t), c);
					break;
				case 3:
					f = c[0];
					break;
				case 1:
					f = Y.toValue(t)[gn(i)](...c);
					break;
			}
			return mn(o, a, f);
		}));
	}, vn = (e) => e ? (e = gn(e), Y.toHandle(globalThis[e])) : Y.toHandle(globalThis), yn = (e) => {
		e > 9 && (J[e + 1] += 1);
	}, bn = (e, t, n, r, i) => un[e](t, n, r, i), xn = (e) => {
		Pe(Y.toValue(e)), Mt(e);
	}, Sn = (e, t, n, r) => {
		var i = (/* @__PURE__ */ new Date()).getFullYear(), a = new Date(i, 0, 1), o = new Date(i, 6, 1), s = a.getTimezoneOffset(), c = o.getTimezoneOffset(), l = Math.max(s, c);
		w[e >> 2] = l * 60, x[t >> 2] = Number(s != c);
		var u = (e) => {
			var t = e >= 0 ? "-" : "+", n = Math.abs(e);
			return `UTC${t}${String(Math.floor(n / 60)).padStart(2, "0")}${String(n % 60).padStart(2, "0")}`;
		}, d = u(s), f = u(c);
		c < s ? (X(d, n, 17), X(f, r, 17)) : (X(d, r, 17), X(f, n, 17));
	}, Cn = () => 2147483648, wn = (e, t) => Math.ceil(e / t) * t, Tn = (e) => {
		var t = (e - Xn.buffer.byteLength + 65535) / 65536 | 0;
		try {
			return Xn.grow(t), ne(), 1;
		} catch {}
	}, En = (e) => {
		var t = T.length;
		e >>>= 0;
		var n = Cn();
		if (e > n) return !1;
		for (var r = 1; r <= 4; r *= 2) {
			var i = t * (1 + .2 / r);
			if (i = Math.min(i, e + 100663296), Tn(Math.min(n, wn(Math.max(e, i), 65536)))) return !0;
		}
		return !1;
	}, Dn = {}, On = () => c || "./this.program", Z = () => {
		if (!Z.strings) {
			var e, t, n = {
				USER: "web_user",
				LOGNAME: "web_user",
				PATH: "/",
				PWD: "/",
				HOME: "/home/web_user",
				LANG: ((e = (t = globalThis.navigator) == null ? void 0 : t.language) == null ? "C" : e).replace("-", "_") + ".UTF-8",
				_: On()
			};
			for (var r in Dn) Dn[r] === void 0 ? delete n[r] : n[r] = Dn[r];
			var i = [];
			for (var r in n) i.push(`${r}=${n[r]}`);
			Z.strings = i;
		}
		return Z.strings;
	}, kn = (e, t) => {
		var n = 0, r = 0;
		for (var i of Z()) {
			var a = t + n;
			w[e + r >> 2] = a, n += X(i, a, Infinity) + 1, r += 4;
		}
		return 0;
	}, An = (e, t) => {
		var n = Z();
		w[e >> 2] = n.length;
		var r = 0;
		for (var i of n) r += Kt(i) + 1;
		return w[t >> 2] = r, 0;
	}, jn = (e) => 52, Mn = (e, t) => t + 2097152 >>> 0 < 4194305 - !!e ? (e >>> 0) + t * 4294967296 : NaN;
	function Nn(e, t, n, r, i) {
		return Mn(t, n), 70;
	}
	var Pn = [
		null,
		[],
		[]
	], Fn = (e, t) => {
		var n = Pn[e];
		t === 0 || t === 10 ? ((e === 1 ? m : h)(Yt(n)), n.length = 0) : n.push(t);
	}, In = (e, t, n, r) => {
		for (var i = 0, a = 0; a < n; a++) {
			var o = w[t >> 2], s = w[t + 4 >> 2];
			t += 8;
			for (var c = 0; c < s; c++) Fn(e, T[o + c]);
			i += s;
		}
		return w[r >> 2] = i, 0;
	}, Ln = (e) => e;
	if (Je(), mt(), i.noExitRuntime && i.noExitRuntime, i.print && (m = i.print), i.printErr && (h = i.printErr), i.wasmBinary && (g = i.wasmBinary), i.arguments && i.arguments, i.thisProgram && (c = i.thisProgram), i.preInit) for (typeof i.preInit == "function" && (i.preInit = [i.preInit]); i.preInit.length > 0;) i.preInit.shift()();
	var Rn, Q, zn, Bn, $, Vn, Hn, Un, Wn, Gn, Kn, qn, Jn, Yn, Xn, Zn;
	function Qn(e) {
		Rn = e.Ca, Q = i._free = e.Da, zn = i._malloc = e.Fa, Bn = e.Ga, $ = e.Ha, Vn = e.Ia, Hn = e.Ja, Un = e.Ka, Wn = e.La, Gn = e.Ma, Kn = e.Na, G.viijii = e.Oa, qn = G.viijjijjjjjj = e.Pa, Jn = G.iiijj = e.Qa, G.jiji = e.Ra, Yn = G.jiiii = e.Sa, G.iiiiij = e.Ta, G.iiiiijj = e.Ua, G.iiiiiijj = e.Va, Xn = e.Aa, Zn = e.Ea;
	}
	var $n = {
		s: Se,
		x: Ce,
		a: Ee,
		i: De,
		m: Oe,
		R: ke,
		p: Ae,
		ga: je,
		d: Me,
		ca: Ne,
		xa: Ie,
		ba: Le,
		ra: ze,
		va: wt,
		ua: Ot,
		H: At,
		pa: Pt,
		X: It,
		Y: Lt,
		A: zt,
		ta: Vt,
		u: Ht,
		wa: Wt,
		qa: Zt,
		S: on,
		I: sn,
		ya: cn,
		sa: ln,
		O: _n,
		za: Mt,
		E: vn,
		T: yn,
		N: bn,
		la: xn,
		da: Sn,
		ha: En,
		ea: kn,
		fa: An,
		ia: jn,
		_: Nn,
		V: In,
		na: xr,
		M: wr,
		C: Ar,
		P: sr,
		U: Mr,
		q: br,
		b: nr,
		F: Cr,
		ka: Or,
		c: ir,
		ja: kr,
		g: or,
		j: pr,
		r: hr,
		Q: Sr,
		t: _r,
		G: vr,
		B: yr,
		K: Nr,
		$: Ir,
		Z: Lr,
		f: cr,
		l: er,
		e: rr,
		W: Tr,
		h: ar,
		L: jr,
		k: tr,
		ma: Er,
		o: gr,
		y: ur,
		v: mr,
		D: fr,
		w: Dr,
		n: lr,
		J: Pr,
		oa: dr,
		aa: Fr,
		z: Ln
	};
	function er(e, t) {
		var n = D();
		try {
			K(e)(t);
		} catch (e) {
			if (E(n), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function tr(e, t, n, r, i) {
		var a = D();
		try {
			K(e)(t, n, r, i);
		} catch (e) {
			if (E(a), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function nr(e, t) {
		var n = D();
		try {
			return K(e)(t);
		} catch (e) {
			if (E(n), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function rr(e, t, n) {
		var r = D();
		try {
			K(e)(t, n);
		} catch (e) {
			if (E(r), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function ir(e, t, n) {
		var r = D();
		try {
			return K(e)(t, n);
		} catch (e) {
			if (E(r), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function ar(e, t, n, r) {
		var i = D();
		try {
			K(e)(t, n, r);
		} catch (e) {
			if (E(i), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function or(e, t, n, r) {
		var i = D();
		try {
			return K(e)(t, n, r);
		} catch (e) {
			if (E(i), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function sr(e, t, n, r, i, a) {
		var o = D();
		try {
			return K(e)(t, n, r, i, a);
		} catch (e) {
			if (E(o), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function cr(e) {
		var t = D();
		try {
			K(e)();
		} catch (e) {
			if (E(t), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function lr(e, t, n, r, i, a, o, s, c, l, u) {
		var d = D();
		try {
			K(e)(t, n, r, i, a, o, s, c, l, u);
		} catch (e) {
			if (E(d), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function ur(e, t, n, r, i, a, o) {
		var s = D();
		try {
			K(e)(t, n, r, i, a, o);
		} catch (e) {
			if (E(s), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function dr(e, t, n, r, i, a, o, s, c, l, u, d, f, p, m, h, g) {
		var _ = D();
		try {
			K(e)(t, n, r, i, a, o, s, c, l, u, d, f, p, m, h, g);
		} catch (e) {
			if (E(_), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function fr(e, t, n, r, i, a, o, s, c) {
		var l = D();
		try {
			K(e)(t, n, r, i, a, o, s, c);
		} catch (e) {
			if (E(l), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function pr(e, t, n, r, i) {
		var a = D();
		try {
			return K(e)(t, n, r, i);
		} catch (e) {
			if (E(a), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function mr(e, t, n, r, i, a, o, s) {
		var c = D();
		try {
			K(e)(t, n, r, i, a, o, s);
		} catch (e) {
			if (E(c), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function hr(e, t, n, r, i, a) {
		var o = D();
		try {
			return K(e)(t, n, r, i, a);
		} catch (e) {
			if (E(o), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function gr(e, t, n, r, i, a) {
		var o = D();
		try {
			K(e)(t, n, r, i, a);
		} catch (e) {
			if (E(o), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function _r(e, t, n, r, i, a, o) {
		var s = D();
		try {
			return K(e)(t, n, r, i, a, o);
		} catch (e) {
			if (E(s), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function vr(e, t, n, r, i, a, o, s) {
		var c = D();
		try {
			return K(e)(t, n, r, i, a, o, s);
		} catch (e) {
			if (E(c), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function yr(e, t, n, r, i, a, o, s, c) {
		var l = D();
		try {
			return K(e)(t, n, r, i, a, o, s, c);
		} catch (e) {
			if (E(l), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function br(e) {
		var t = D();
		try {
			return K(e)();
		} catch (e) {
			if (E(t), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function xr(e, t, n) {
		var r = D();
		try {
			return K(e)(t, n);
		} catch (e) {
			if (E(r), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Sr(e, t, n, r, i, a, o) {
		var s = D();
		try {
			return K(e)(t, n, r, i, a, o);
		} catch (e) {
			if (E(s), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Cr(e, t, n, r) {
		var i = D();
		try {
			return K(e)(t, n, r);
		} catch (e) {
			if (E(i), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function wr(e, t, n, r) {
		var i = D();
		try {
			return K(e)(t, n, r);
		} catch (e) {
			if (E(i), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Tr(e, t, n, r, i, a, o, s, c) {
		var l = D();
		try {
			K(e)(t, n, r, i, a, o, s, c);
		} catch (e) {
			if (E(l), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Er(e, t, n, r, i, a, o, s) {
		var c = D();
		try {
			K(e)(t, n, r, i, a, o, s);
		} catch (e) {
			if (E(c), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Dr(e, t, n, r, i, a, o, s, c, l) {
		var u = D();
		try {
			K(e)(t, n, r, i, a, o, s, c, l);
		} catch (e) {
			if (E(u), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Or(e, t, n) {
		var r = D();
		try {
			return K(e)(t, n);
		} catch (e) {
			if (E(r), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function kr(e, t, n, r, i) {
		var a = D();
		try {
			return K(e)(t, n, r, i);
		} catch (e) {
			if (E(a), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Ar(e, t, n, r, i, a) {
		var o = D();
		try {
			return K(e)(t, n, r, i, a);
		} catch (e) {
			if (E(o), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function jr(e, t, n, r, i, a, o) {
		var s = D();
		try {
			K(e)(t, n, r, i, a, o);
		} catch (e) {
			if (E(s), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Mr(e, t, n, r) {
		var i = D();
		try {
			return K(e)(t, n, r);
		} catch (e) {
			if (E(i), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Nr(e, t, n, r, i, a, o, s, c, l, u, d) {
		var f = D();
		try {
			return K(e)(t, n, r, i, a, o, s, c, l, u, d);
		} catch (e) {
			if (E(f), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Pr(e, t, n, r, i, a, o, s, c, l, u, d, f, p, m, h) {
		var g = D();
		try {
			K(e)(t, n, r, i, a, o, s, c, l, u, d, f, p, m, h);
		} catch (e) {
			if (E(g), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Fr(e, t, n, r, i, a, o, s, c, l, u, d, f, p, m, h, g, _, v, ee) {
		var te = D();
		try {
			qn(e, t, n, r, i, a, o, s, c, l, u, d, f, p, m, h, g, _, v, ee);
		} catch (e) {
			if (E(te), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Ir(e, t, n, r, i, a, o) {
		var s = D();
		try {
			return Jn(e, t, n, r, i, a, o);
		} catch (e) {
			if (E(s), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Lr(e, t, n, r, i) {
		var a = D();
		try {
			return Yn(e, t, n, r, i);
		} catch (e) {
			if (E(a), e !== e + 0) throw e;
			$(1, 0);
		}
	}
	function Rr() {
		re();
		function e() {
			var e, t;
			i.calledRun = !0, !_ && (ie(), (e = v) == null || e(i), (t = i.onRuntimeInitialized) == null || t.call(i), ae());
		}
		i.setStatus ? (i.setStatus("Running..."), setTimeout(() => {
			setTimeout(() => i.setStatus(""), 1), e();
		}, 1)) : e();
	}
	var zr = await pe();
	return Rr(), t = te ? i : new Promise((e, t) => {
		v = e, ee = t;
	}), t;
}
//#endregion
//#region src/reader/index.ts
function S(e) {
	return m(x, e);
}
function me() {
	return y(x);
}
function he(e) {
	return S({
		overrides: e,
		equalityFn: Object.is,
		fireImmediately: !0
	});
}
function C(e) {
	S({
		overrides: e,
		equalityFn: Object.is,
		fireImmediately: !1
	});
}
async function w(e, t) {
	return le(x, e, t);
}
async function T(e, t) {
	return w(e, t);
}
async function ge(e, t) {
	return w(e, t);
}
var _e = "319c39440f2237a31a4c41d9e72fa053f8e35e618df5aa4eac1264db0f3da008";
//#endregion
export { b as BARCODE_FORMATS, h as BARCODE_HRI_LABELS, pe as BARCODE_META_FORMATS, d as BARCODE_SYMBOLOGIES, p as BINARIZERS, ee as CHARACTER_SETS, se as CONTENT_TYPES, t as CREATABLE_BARCODE_FORMATS, g as EAN_ADD_ON_SYMBOLS, fe as GS1_BARCODE_FORMATS, f as INDUSTRIAL_BARCODE_FORMATS, r as LINEAR_BARCODE_FORMATS, n as MATRIX_BARCODE_FORMATS, l as READABLE_BARCODE_FORMATS, re as RETAIL_BARCODE_FORMATS, ie as TEXT_MODES, ce as ZXING_CPP_COMMIT, _e as ZXING_WASM_SHA256, te as ZXING_WASM_VERSION, e as barcodeFormats, de as binarizers, v as characterSets, ae as contentTypes, oe as defaultReaderOptions, _ as eanAddOnSymbols, ne as encodeFormat, s as encodeFormats, c as formatToLabel, u as formatToSymbology, he as getZXingModule, i as linearBarcodeFormats, a as matrixBarcodeFormats, S as prepareZXingModule, me as purgeZXingModule, w as readBarcodes, ge as readBarcodesFromImageData, T as readBarcodesFromImageFile, C as setZXingModuleOverrides, o as symbologyToFormats, ue as textModes };
